// @rarible/solana-sdk/build/sdk/prepared-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreparedTransaction = void 0;
var tslib_1 = require("tslib");
var transactions_1 = require("../common/transactions");
var PreparedTransaction = /** @class */ (function () {
    function PreparedTransaction(connection, data, signer, logger, onSubmit) {
        this.connection = connection;
        this.data = data;
        this.signer = signer;
        this.logger = logger;
        this.onSubmit = onSubmit;
    }
    PreparedTransaction.prototype.submit = function (commitment) {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var res;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, (0, transactions_1.sendTransactionWithRetry)(this.connection, this.signer, this.data.instructions, this.data.signers, commitment, this.logger)];
                    case 1:
                        res = _b.sent();
                        (_a = this.onSubmit) === null || _a === void 0 ? void 0 : _a.call(this, res);
                        return [2 /*return*/, res];
                }
            });
        });
    };
    return PreparedTransaction;
}());
exports.PreparedTransaction = PreparedTransaction;
