// @rarible/api-client/build/models/OrderData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthOrderOpenSeaV1DataV1HowToCall = exports.EthOrderOpenSeaV1DataV1SaleKind = exports.EthOrderOpenSeaV1DataV1Side = exports.EthOrderOpenSeaV1DataV1FeeMethod = void 0;
var EthOrderOpenSeaV1DataV1FeeMethod;
(function (EthOrderOpenSeaV1DataV1FeeMethod) {
    EthOrderOpenSeaV1DataV1FeeMethod["PROTOCOL_FEE"] = "PROTOCOL_FEE";
    EthOrderOpenSeaV1DataV1FeeMethod["SPLIT_FEE"] = "SPLIT_FEE";
})(EthOrderOpenSeaV1DataV1FeeMethod = exports.EthOrderOpenSeaV1DataV1FeeMethod || (exports.EthOrderOpenSeaV1DataV1FeeMethod = {}));
var EthOrderOpenSeaV1DataV1Side;
(function (EthOrderOpenSeaV1DataV1Side) {
    EthOrderOpenSeaV1DataV1Side["BUY"] = "BUY";
    EthOrderOpenSeaV1DataV1Side["SELL"] = "SELL";
})(EthOrderOpenSeaV1DataV1Side = exports.EthOrderOpenSeaV1DataV1Side || (exports.EthOrderOpenSeaV1DataV1Side = {}));
var EthOrderOpenSeaV1DataV1SaleKind;
(function (EthOrderOpenSeaV1DataV1SaleKind) {
    EthOrderOpenSeaV1DataV1SaleKind["FIXED_PRICE"] = "FIXED_PRICE";
    EthOrderOpenSeaV1DataV1SaleKind["DUTCH_AUCTION"] = "DUTCH_AUCTION";
})(EthOrderOpenSeaV1DataV1SaleKind = exports.EthOrderOpenSeaV1DataV1SaleKind || (exports.EthOrderOpenSeaV1DataV1SaleKind = {}));
var EthOrderOpenSeaV1DataV1HowToCall;
(function (EthOrderOpenSeaV1DataV1HowToCall) {
    EthOrderOpenSeaV1DataV1HowToCall["CALL"] = "CALL";
    EthOrderOpenSeaV1DataV1HowToCall["DELEGATE_CALL"] = "DELEGATE_CALL";
})(EthOrderOpenSeaV1DataV1HowToCall = exports.EthOrderOpenSeaV1DataV1HowToCall || (exports.EthOrderOpenSeaV1DataV1HowToCall = {}));
