// @rarible/api-client/build/models/UserActivityType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserActivityType = void 0;
var UserActivityType;
(function (UserActivityType) {
    UserActivityType["TRANSFER_FROM"] = "TRANSFER_FROM";
    UserActivityType["TRANSFER_TO"] = "TRANSFER_TO";
    UserActivityType["MINT"] = "MINT";
    UserActivityType["BURN"] = "BURN";
    UserActivityType["MAKE_BID"] = "MAKE_BID";
    UserActivityType["GET_BID"] = "GET_BID";
    UserActivityType["LIST"] = "LIST";
    UserActivityType["BUY"] = "BUY";
    UserActivityType["SELL"] = "SELL";
    UserActivityType["CANCEL_LIST"] = "CANCEL_LIST";
    UserActivityType["CANCEL_BID"] = "CANCEL_BID";
    UserActivityType["AUCTION_BID"] = "AUCTION_BID";
    UserActivityType["AUCTION_CREATED"] = "AUCTION_CREATED";
    UserActivityType["AUCTION_CANCEL"] = "AUCTION_CANCEL";
    UserActivityType["AUCTION_FINISHED"] = "AUCTION_FINISHED";
    UserActivityType["AUCTION_STARTED"] = "AUCTION_STARTED";
    UserActivityType["AUCTION_ENDED"] = "AUCTION_ENDED";
})(UserActivityType = exports.UserActivityType || (exports.UserActivityType = {}));
