// @rarible/solana-sdk/build/sdk/collection/verifyCollection/verify-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVerifyCollectionInstructions = void 0;
var tslib_1 = require("tslib");
var mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
var helpers_1 = require("../../../common/helpers");
function getVerifyCollectionInstructions(request) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var metadataAccount, collectionMetadataAccount, collectionMasterEdition, signers, tx;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, helpers_1.getMetadata)(request.mint)];
                case 1:
                    metadataAccount = _a.sent();
                    return [4 /*yield*/, (0, helpers_1.getMetadata)(request.collection)];
                case 2:
                    collectionMetadataAccount = _a.sent();
                    return [4 /*yield*/, (0, helpers_1.getMasterEdition)(request.collection)];
                case 3:
                    collectionMasterEdition = _a.sent();
                    signers = [request.signer];
                    tx = new mpl_token_metadata_1.VerifyCollection({ feePayer: request.signer.publicKey }, {
                        metadata: metadataAccount,
                        collectionAuthority: request.signer.publicKey,
                        collectionMint: request.collection,
                        collectionMetadata: collectionMetadataAccount,
                        collectionMasterEdition: collectionMasterEdition,
                    });
                    return [2 /*return*/, { instructions: tx.instructions, signers: signers }];
            }
        });
    });
}
exports.getVerifyCollectionInstructions = getVerifyCollectionInstructions;
