// @rarible/ethereum-api-client/build/models/AuctionActivityFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuctionActivityFilterByCollectionTypes = exports.AuctionActivityFilterByItemTypes = exports.AuctionActivityFilterByUserTypes = exports.AuctionActivityFilterAllTypes = void 0;
var AuctionActivityFilterAllTypes;
(function (AuctionActivityFilterAllTypes) {
    AuctionActivityFilterAllTypes["CREATED"] = "CREATED";
    AuctionActivityFilterAllTypes["BID"] = "BID";
    AuctionActivityFilterAllTypes["CANCEL"] = "CANCEL";
    AuctionActivityFilterAllTypes["FINISHED"] = "FINISHED";
    AuctionActivityFilterAllTypes["STARTED"] = "STARTED";
    AuctionActivityFilterAllTypes["ENDED"] = "ENDED";
})(AuctionActivityFilterAllTypes = exports.AuctionActivityFilterAllTypes || (exports.AuctionActivityFilterAllTypes = {}));
var AuctionActivityFilterByUserTypes;
(function (AuctionActivityFilterByUserTypes) {
    AuctionActivityFilterByUserTypes["CREATED"] = "CREATED";
    AuctionActivityFilterByUserTypes["BID"] = "BID";
    AuctionActivityFilterByUserTypes["CANCEL"] = "CANCEL";
    AuctionActivityFilterByUserTypes["FINISHED"] = "FINISHED";
    AuctionActivityFilterByUserTypes["STARTED"] = "STARTED";
    AuctionActivityFilterByUserTypes["ENDED"] = "ENDED";
})(AuctionActivityFilterByUserTypes = exports.AuctionActivityFilterByUserTypes || (exports.AuctionActivityFilterByUserTypes = {}));
var AuctionActivityFilterByItemTypes;
(function (AuctionActivityFilterByItemTypes) {
    AuctionActivityFilterByItemTypes["CREATED"] = "CREATED";
    AuctionActivityFilterByItemTypes["BID"] = "BID";
    AuctionActivityFilterByItemTypes["CANCEL"] = "CANCEL";
    AuctionActivityFilterByItemTypes["FINISHED"] = "FINISHED";
    AuctionActivityFilterByItemTypes["STARTED"] = "STARTED";
    AuctionActivityFilterByItemTypes["ENDED"] = "ENDED";
})(AuctionActivityFilterByItemTypes = exports.AuctionActivityFilterByItemTypes || (exports.AuctionActivityFilterByItemTypes = {}));
var AuctionActivityFilterByCollectionTypes;
(function (AuctionActivityFilterByCollectionTypes) {
    AuctionActivityFilterByCollectionTypes["CREATED"] = "CREATED";
    AuctionActivityFilterByCollectionTypes["BID"] = "BID";
    AuctionActivityFilterByCollectionTypes["CANCEL"] = "CANCEL";
    AuctionActivityFilterByCollectionTypes["FINISHED"] = "FINISHED";
    AuctionActivityFilterByCollectionTypes["STARTED"] = "STARTED";
    AuctionActivityFilterByCollectionTypes["ENDED"] = "ENDED";
})(AuctionActivityFilterByCollectionTypes = exports.AuctionActivityFilterByCollectionTypes || (exports.AuctionActivityFilterByCollectionTypes = {}));
