// @rarible/ethereum-api-client/build/models/NftActivityFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterByCollectionTypes = exports.NftActivityFilterByItemTypes = exports.NftActivityFilterByUserTypes = exports.NftActivityFilterAllTypes = void 0;
var NftActivityFilterAllTypes;
(function (NftActivityFilterAllTypes) {
    NftActivityFilterAllTypes["TRANSFER"] = "TRANSFER";
    NftActivityFilterAllTypes["MINT"] = "MINT";
    NftActivityFilterAllTypes["BURN"] = "BURN";
})(NftActivityFilterAllTypes = exports.NftActivityFilterAllTypes || (exports.NftActivityFilterAllTypes = {}));
var NftActivityFilterByUserTypes;
(function (NftActivityFilterByUserTypes) {
    NftActivityFilterByUserTypes["TRANSFER_FROM"] = "TRANSFER_FROM";
    NftActivityFilterByUserTypes["TRANSFER_TO"] = "TRANSFER_TO";
    NftActivityFilterByUserTypes["MINT"] = "MINT";
    NftActivityFilterByUserTypes["BURN"] = "BURN";
})(NftActivityFilterByUserTypes = exports.NftActivityFilterByUserTypes || (exports.NftActivityFilterByUserTypes = {}));
var NftActivityFilterByItemTypes;
(function (NftActivityFilterByItemTypes) {
    NftActivityFilterByItemTypes["TRANSFER"] = "TRANSFER";
    NftActivityFilterByItemTypes["MINT"] = "MINT";
    NftActivityFilterByItemTypes["BURN"] = "BURN";
})(NftActivityFilterByItemTypes = exports.NftActivityFilterByItemTypes || (exports.NftActivityFilterByItemTypes = {}));
var NftActivityFilterByCollectionTypes;
(function (NftActivityFilterByCollectionTypes) {
    NftActivityFilterByCollectionTypes["TRANSFER"] = "TRANSFER";
    NftActivityFilterByCollectionTypes["MINT"] = "MINT";
    NftActivityFilterByCollectionTypes["BURN"] = "BURN";
})(NftActivityFilterByCollectionTypes = exports.NftActivityFilterByCollectionTypes || (exports.NftActivityFilterByCollectionTypes = {}));
