// @rarible/ethereum-api-client/build/models/NftCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftCollectionFeatures = exports.NftCollectionStatus = exports.NftCollectionType = void 0;
var NftCollectionType;
(function (NftCollectionType) {
    NftCollectionType["ERC721"] = "ERC721";
    NftCollectionType["ERC1155"] = "ERC1155";
    NftCollectionType["CRYPTO_PUNKS"] = "CRYPTO_PUNKS";
})(NftCollectionType = exports.NftCollectionType || (exports.NftCollectionType = {}));
var NftCollectionStatus;
(function (NftCollectionStatus) {
    NftCollectionStatus["PENDING"] = "PENDING";
    NftCollectionStatus["ERROR"] = "ERROR";
    NftCollectionStatus["CONFIRMED"] = "CONFIRMED";
})(NftCollectionStatus = exports.NftCollectionStatus || (exports.NftCollectionStatus = {}));
var NftCollectionFeatures;
(function (NftCollectionFeatures) {
    NftCollectionFeatures["APPROVE_FOR_ALL"] = "APPROVE_FOR_ALL";
    NftCollectionFeatures["SET_URI_PREFIX"] = "SET_URI_PREFIX";
    NftCollectionFeatures["BURN"] = "BURN";
    NftCollectionFeatures["MINT_WITH_ADDRESS"] = "MINT_WITH_ADDRESS";
    NftCollectionFeatures["SECONDARY_SALE_FEES"] = "SECONDARY_SALE_FEES";
    NftCollectionFeatures["MINT_AND_TRANSFER"] = "MINT_AND_TRANSFER";
})(NftCollectionFeatures = exports.NftCollectionFeatures || (exports.NftCollectionFeatures = {}));
