// @rarible/ethereum-api-client/build/models/OrderData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderOpenSeaV1DataV1HowToCall = exports.OrderOpenSeaV1DataV1SaleKind = exports.OrderOpenSeaV1DataV1Side = exports.OrderOpenSeaV1DataV1FeeMethod = void 0;
var OrderOpenSeaV1DataV1FeeMethod;
(function (OrderOpenSeaV1DataV1FeeMethod) {
    OrderOpenSeaV1DataV1FeeMethod["PROTOCOL_FEE"] = "PROTOCOL_FEE";
    OrderOpenSeaV1DataV1FeeMethod["SPLIT_FEE"] = "SPLIT_FEE";
})(OrderOpenSeaV1DataV1FeeMethod = exports.OrderOpenSeaV1DataV1FeeMethod || (exports.OrderOpenSeaV1DataV1FeeMethod = {}));
var OrderOpenSeaV1DataV1Side;
(function (OrderOpenSeaV1DataV1Side) {
    OrderOpenSeaV1DataV1Side["BUY"] = "BUY";
    OrderOpenSeaV1DataV1Side["SELL"] = "SELL";
})(OrderOpenSeaV1DataV1Side = exports.OrderOpenSeaV1DataV1Side || (exports.OrderOpenSeaV1DataV1Side = {}));
var OrderOpenSeaV1DataV1SaleKind;
(function (OrderOpenSeaV1DataV1SaleKind) {
    OrderOpenSeaV1DataV1SaleKind["FIXED_PRICE"] = "FIXED_PRICE";
    OrderOpenSeaV1DataV1SaleKind["DUTCH_AUCTION"] = "DUTCH_AUCTION";
})(OrderOpenSeaV1DataV1SaleKind = exports.OrderOpenSeaV1DataV1SaleKind || (exports.OrderOpenSeaV1DataV1SaleKind = {}));
var OrderOpenSeaV1DataV1HowToCall;
(function (OrderOpenSeaV1DataV1HowToCall) {
    OrderOpenSeaV1DataV1HowToCall["CALL"] = "CALL";
    OrderOpenSeaV1DataV1HowToCall["DELEGATE_CALL"] = "DELEGATE_CALL";
})(OrderOpenSeaV1DataV1HowToCall = exports.OrderOpenSeaV1DataV1HowToCall || (exports.OrderOpenSeaV1DataV1HowToCall = {}));
