// @rarible/ethereum-api-client/build/models/OrderActivityFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterByCollectionTypes = exports.OrderActivityFilterByItemTypes = exports.OrderActivityFilterByUserTypes = exports.OrderActivityFilterAllTypes = void 0;
var OrderActivityFilterAllTypes;
(function (OrderActivityFilterAllTypes) {
    OrderActivityFilterAllTypes["BID"] = "BID";
    OrderActivityFilterAllTypes["LIST"] = "LIST";
    OrderActivityFilterAllTypes["MATCH"] = "MATCH";
    OrderActivityFilterAllTypes["CANCEL_BID"] = "CANCEL_BID";
    OrderActivityFilterAllTypes["CANCEL_LIST"] = "CANCEL_LIST";
})(OrderActivityFilterAllTypes = exports.OrderActivityFilterAllTypes || (exports.OrderActivityFilterAllTypes = {}));
var OrderActivityFilterByUserTypes;
(function (OrderActivityFilterByUserTypes) {
    OrderActivityFilterByUserTypes["MAKE_BID"] = "MAKE_BID";
    OrderActivityFilterByUserTypes["GET_BID"] = "GET_BID";
    OrderActivityFilterByUserTypes["LIST"] = "LIST";
    OrderActivityFilterByUserTypes["BUY"] = "BUY";
    OrderActivityFilterByUserTypes["SELL"] = "SELL";
    OrderActivityFilterByUserTypes["CANCEL_BID"] = "CANCEL_BID";
    OrderActivityFilterByUserTypes["CANCEL_LIST"] = "CANCEL_LIST";
})(OrderActivityFilterByUserTypes = exports.OrderActivityFilterByUserTypes || (exports.OrderActivityFilterByUserTypes = {}));
var OrderActivityFilterByItemTypes;
(function (OrderActivityFilterByItemTypes) {
    OrderActivityFilterByItemTypes["BID"] = "BID";
    OrderActivityFilterByItemTypes["LIST"] = "LIST";
    OrderActivityFilterByItemTypes["MATCH"] = "MATCH";
    OrderActivityFilterByItemTypes["CANCEL_BID"] = "CANCEL_BID";
    OrderActivityFilterByItemTypes["CANCEL_LIST"] = "CANCEL_LIST";
})(OrderActivityFilterByItemTypes = exports.OrderActivityFilterByItemTypes || (exports.OrderActivityFilterByItemTypes = {}));
var OrderActivityFilterByCollectionTypes;
(function (OrderActivityFilterByCollectionTypes) {
    OrderActivityFilterByCollectionTypes["BID"] = "BID";
    OrderActivityFilterByCollectionTypes["LIST"] = "LIST";
    OrderActivityFilterByCollectionTypes["MATCH"] = "MATCH";
    OrderActivityFilterByCollectionTypes["CANCEL_BID"] = "CANCEL_BID";
    OrderActivityFilterByCollectionTypes["CANCEL_LIST"] = "CANCEL_LIST";
})(OrderActivityFilterByCollectionTypes = exports.OrderActivityFilterByCollectionTypes || (exports.OrderActivityFilterByCollectionTypes = {}));
