// @rarible/tezos-sdk/dist/order/cancel.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancel = exports.cancel_arg = void 0;
var tezos_common_1 = require("@rarible/tezos-common");
var sign_order_1 = require("./sign-order");
var wrapper_1 = require("./wrapper");
var bignumber_js_1 = require("bignumber.js");
function cancel_arg(provider, order) {
    return __awaiter(this, void 0, void 0, function () {
        var parameter;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, sign_order_1.order_to_struct)(provider, order)];
                case 1:
                    parameter = _a.sent();
                    return [2 /*return*/, {
                            destination: provider.config.exchange,
                            entrypoint: "cancel",
                            parameter: parameter,
                            amount: new bignumber_js_1.default(0)
                        }];
            }
        });
    });
}
exports.cancel_arg = cancel_arg;
function cancel(provider, order, unwrap, unwrap_amount) {
    if (unwrap === void 0) { unwrap = false; }
    return __awaiter(this, void 0, void 0, function () {
        var arg, _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0: return [4 /*yield*/, cancel_arg(provider, order)];
                case 1:
                    arg = [_c.sent()];
                    if (!(unwrap && order.make.asset_type.asset_class == "FT" && order.make.asset_type.contract == provider.config.wrapper && order.make.asset_type.token_id != undefined && order.make.asset_type.token_id.isZero())) return [3 /*break*/, 3];
                    _b = (_a = arg).concat;
                    return [4 /*yield*/, (0, wrapper_1.unwrap_arg)(provider, unwrap_amount || order.make.value)];
                case 2:
                    arg = _b.apply(_a, [_c.sent()]);
                    _c.label = 3;
                case 3: return [2 /*return*/, (0, tezos_common_1.send_batch)(provider, arg)];
            }
        });
    });
}
exports.cancel = cancel;
