// @rarible/flow-api-client/build/apis/FlowNftCryptoControllerApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * OpenAPI definition
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: v1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowNftCryptoControllerApi = void 0;
var tslib_1 = require("tslib");
var runtime = (0, tslib_1.__importStar)(require("../runtime"));
/**
 * no description
 */
var FlowNftCryptoControllerApi = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(FlowNftCryptoControllerApi, _super);
    function FlowNftCryptoControllerApi() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     */
    FlowNftCryptoControllerApi.prototype.verifySignatureRaw = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var queryParameters, headerParameters, response;
            var _a, _b;
            return (0, tslib_1.__generator)(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (requestParameters.publicKey === null || requestParameters.publicKey === undefined) {
                            throw new runtime.RequiredError('publicKey', 'Required parameter requestParameters.publicKey was null or undefined when calling verifySignature.');
                        }
                        if (requestParameters.signerAddress === null || requestParameters.signerAddress === undefined) {
                            throw new runtime.RequiredError('signerAddress', 'Required parameter requestParameters.signerAddress was null or undefined when calling verifySignature.');
                        }
                        if (requestParameters.signature === null || requestParameters.signature === undefined) {
                            throw new runtime.RequiredError('signature', 'Required parameter requestParameters.signature was null or undefined when calling verifySignature.');
                        }
                        if (requestParameters.message === null || requestParameters.message === undefined) {
                            throw new runtime.RequiredError('message', 'Required parameter requestParameters.message was null or undefined when calling verifySignature.');
                        }
                        queryParameters = {};
                        if (requestParameters.publicKey !== undefined) {
                            queryParameters['publicKey'] = requestParameters.publicKey;
                        }
                        if (requestParameters.signerAddress !== undefined) {
                            queryParameters['signerAddress'] = requestParameters.signerAddress;
                        }
                        if (requestParameters.signature !== undefined) {
                            queryParameters['signature'] = requestParameters.signature;
                        }
                        if (requestParameters.message !== undefined) {
                            queryParameters['message'] = requestParameters.message;
                        }
                        headerParameters = {};
                        return [4 /*yield*/, this.request({
                                path: "/v0.1/crypto/verify",
                                method: 'GET',
                                headers: headerParameters,
                                query: queryParameters,
                            })];
                    case 1:
                        response = _c.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 3];
                        _a = {
                            status: 200
                        };
                        return [4 /*yield*/, response.json()];
                    case 2: return [2 /*return*/, (_a.value = _c.sent(),
                            _a)];
                    case 3:
                        if (!(response.status === 500)) return [3 /*break*/, 5];
                        _b = {
                            status: 500
                        };
                        return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, (_b.value = _c.sent(),
                            _b)];
                    case 5: throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftCryptoControllerApi.prototype.verifySignature = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.verifySignatureRaw(requestParameters)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 200) {
                            return [2 /*return*/, response.value];
                        }
                        throw response;
                }
            });
        });
    };
    return FlowNftCryptoControllerApi;
}(runtime.BaseAPI));
exports.FlowNftCryptoControllerApi = FlowNftCryptoControllerApi;
