// @rarible/flow-api-client/build/apis/FlowNftOrderActivityControllerApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * OpenAPI definition
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: v1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetNftOrderAllActivitiesSortEnum = exports.GetNftOrderAllActivitiesTypeEnum = exports.GetNftOrderActivitiesByUserSortEnum = exports.GetNftOrderActivitiesByUserTypeEnum = exports.GetNftOrderActivitiesByItemSortEnum = exports.GetNftOrderActivitiesByItemTypeEnum = exports.GetNftOrderActivitiesByCollectionSortEnum = exports.GetNftOrderActivitiesByCollectionTypeEnum = exports.FlowNftOrderActivityControllerApi = void 0;
var tslib_1 = require("tslib");
var runtime = (0, tslib_1.__importStar)(require("../runtime"));
/**
 * no description
 */
var FlowNftOrderActivityControllerApi = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(FlowNftOrderActivityControllerApi, _super);
    function FlowNftOrderActivityControllerApi() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByCollectionRaw = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var queryParameters, headerParameters, response;
            var _a, _b, _c;
            return (0, tslib_1.__generator)(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (requestParameters.type === null || requestParameters.type === undefined) {
                            throw new runtime.RequiredError('type', 'Required parameter requestParameters.type was null or undefined when calling getNftOrderActivitiesByCollection.');
                        }
                        if (requestParameters.collection === null || requestParameters.collection === undefined) {
                            throw new runtime.RequiredError('collection', 'Required parameter requestParameters.collection was null or undefined when calling getNftOrderActivitiesByCollection.');
                        }
                        queryParameters = {};
                        if (requestParameters.type) {
                            queryParameters['type'] = requestParameters.type;
                        }
                        if (requestParameters.collection !== undefined) {
                            queryParameters['collection'] = requestParameters.collection;
                        }
                        if (requestParameters.continuation !== undefined) {
                            queryParameters['continuation'] = requestParameters.continuation;
                        }
                        if (requestParameters.size !== undefined) {
                            queryParameters['size'] = requestParameters.size;
                        }
                        if (requestParameters.sort !== undefined) {
                            queryParameters['sort'] = requestParameters.sort;
                        }
                        headerParameters = {};
                        return [4 /*yield*/, this.request({
                                path: "/v0.1/order/activities/byCollection",
                                method: 'GET',
                                headers: headerParameters,
                                query: queryParameters,
                            })];
                    case 1:
                        response = _d.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 3];
                        _a = {
                            status: 200
                        };
                        return [4 /*yield*/, response.json()];
                    case 2: return [2 /*return*/, (_a.value = _d.sent(),
                            _a)];
                    case 3:
                        if (!(response.status === 400)) return [3 /*break*/, 5];
                        _b = {
                            status: 400
                        };
                        return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, (_b.value = _d.sent(),
                            _b)];
                    case 5:
                        if (!(response.status === 500)) return [3 /*break*/, 7];
                        _c = {
                            status: 500
                        };
                        return [4 /*yield*/, response.json()];
                    case 6: return [2 /*return*/, (_c.value = _d.sent(),
                            _c)];
                    case 7: throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByCollection = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getNftOrderActivitiesByCollectionRaw(requestParameters)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 200) {
                            return [2 /*return*/, response.value];
                        }
                        throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByItemRaw = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var queryParameters, headerParameters, response;
            var _a, _b, _c;
            return (0, tslib_1.__generator)(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (requestParameters.type === null || requestParameters.type === undefined) {
                            throw new runtime.RequiredError('type', 'Required parameter requestParameters.type was null or undefined when calling getNftOrderActivitiesByItem.');
                        }
                        if (requestParameters.contract === null || requestParameters.contract === undefined) {
                            throw new runtime.RequiredError('contract', 'Required parameter requestParameters.contract was null or undefined when calling getNftOrderActivitiesByItem.');
                        }
                        if (requestParameters.tokenId === null || requestParameters.tokenId === undefined) {
                            throw new runtime.RequiredError('tokenId', 'Required parameter requestParameters.tokenId was null or undefined when calling getNftOrderActivitiesByItem.');
                        }
                        queryParameters = {};
                        if (requestParameters.type) {
                            queryParameters['type'] = requestParameters.type;
                        }
                        if (requestParameters.contract !== undefined) {
                            queryParameters['contract'] = requestParameters.contract;
                        }
                        if (requestParameters.tokenId !== undefined) {
                            queryParameters['tokenId'] = requestParameters.tokenId;
                        }
                        if (requestParameters.continuation !== undefined) {
                            queryParameters['continuation'] = requestParameters.continuation;
                        }
                        if (requestParameters.size !== undefined) {
                            queryParameters['size'] = requestParameters.size;
                        }
                        if (requestParameters.sort !== undefined) {
                            queryParameters['sort'] = requestParameters.sort;
                        }
                        headerParameters = {};
                        return [4 /*yield*/, this.request({
                                path: "/v0.1/order/activities/byItem",
                                method: 'GET',
                                headers: headerParameters,
                                query: queryParameters,
                            })];
                    case 1:
                        response = _d.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 3];
                        _a = {
                            status: 200
                        };
                        return [4 /*yield*/, response.json()];
                    case 2: return [2 /*return*/, (_a.value = _d.sent(),
                            _a)];
                    case 3:
                        if (!(response.status === 400)) return [3 /*break*/, 5];
                        _b = {
                            status: 400
                        };
                        return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, (_b.value = _d.sent(),
                            _b)];
                    case 5:
                        if (!(response.status === 500)) return [3 /*break*/, 7];
                        _c = {
                            status: 500
                        };
                        return [4 /*yield*/, response.json()];
                    case 6: return [2 /*return*/, (_c.value = _d.sent(),
                            _c)];
                    case 7: throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByItem = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getNftOrderActivitiesByItemRaw(requestParameters)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 200) {
                            return [2 /*return*/, response.value];
                        }
                        throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByUserRaw = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var queryParameters, headerParameters, response;
            var _a, _b, _c;
            return (0, tslib_1.__generator)(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (requestParameters.type === null || requestParameters.type === undefined) {
                            throw new runtime.RequiredError('type', 'Required parameter requestParameters.type was null or undefined when calling getNftOrderActivitiesByUser.');
                        }
                        if (requestParameters.user === null || requestParameters.user === undefined) {
                            throw new runtime.RequiredError('user', 'Required parameter requestParameters.user was null or undefined when calling getNftOrderActivitiesByUser.');
                        }
                        queryParameters = {};
                        if (requestParameters.type) {
                            queryParameters['type'] = requestParameters.type;
                        }
                        if (requestParameters.user) {
                            queryParameters['user'] = requestParameters.user;
                        }
                        if (requestParameters.from !== undefined) {
                            queryParameters['from'] = requestParameters.from;
                        }
                        if (requestParameters.to !== undefined) {
                            queryParameters['to'] = requestParameters.to;
                        }
                        if (requestParameters.continuation !== undefined) {
                            queryParameters['continuation'] = requestParameters.continuation;
                        }
                        if (requestParameters.size !== undefined) {
                            queryParameters['size'] = requestParameters.size;
                        }
                        if (requestParameters.sort !== undefined) {
                            queryParameters['sort'] = requestParameters.sort;
                        }
                        headerParameters = {};
                        return [4 /*yield*/, this.request({
                                path: "/v0.1/order/activities/byUser",
                                method: 'GET',
                                headers: headerParameters,
                                query: queryParameters,
                            })];
                    case 1:
                        response = _d.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 3];
                        _a = {
                            status: 200
                        };
                        return [4 /*yield*/, response.json()];
                    case 2: return [2 /*return*/, (_a.value = _d.sent(),
                            _a)];
                    case 3:
                        if (!(response.status === 400)) return [3 /*break*/, 5];
                        _b = {
                            status: 400
                        };
                        return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, (_b.value = _d.sent(),
                            _b)];
                    case 5:
                        if (!(response.status === 500)) return [3 /*break*/, 7];
                        _c = {
                            status: 500
                        };
                        return [4 /*yield*/, response.json()];
                    case 6: return [2 /*return*/, (_c.value = _d.sent(),
                            _c)];
                    case 7: throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderActivitiesByUser = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getNftOrderActivitiesByUserRaw(requestParameters)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 200) {
                            return [2 /*return*/, response.value];
                        }
                        throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderAllActivitiesRaw = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var queryParameters, headerParameters, response;
            var _a, _b, _c;
            return (0, tslib_1.__generator)(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (requestParameters.type === null || requestParameters.type === undefined) {
                            throw new runtime.RequiredError('type', 'Required parameter requestParameters.type was null or undefined when calling getNftOrderAllActivities.');
                        }
                        queryParameters = {};
                        if (requestParameters.type) {
                            queryParameters['type'] = requestParameters.type;
                        }
                        if (requestParameters.continuation !== undefined) {
                            queryParameters['continuation'] = requestParameters.continuation;
                        }
                        if (requestParameters.size !== undefined) {
                            queryParameters['size'] = requestParameters.size;
                        }
                        if (requestParameters.sort !== undefined) {
                            queryParameters['sort'] = requestParameters.sort;
                        }
                        headerParameters = {};
                        return [4 /*yield*/, this.request({
                                path: "/v0.1/order/activities/all",
                                method: 'GET',
                                headers: headerParameters,
                                query: queryParameters,
                            })];
                    case 1:
                        response = _d.sent();
                        if (!(response.status === 200)) return [3 /*break*/, 3];
                        _a = {
                            status: 200
                        };
                        return [4 /*yield*/, response.json()];
                    case 2: return [2 /*return*/, (_a.value = _d.sent(),
                            _a)];
                    case 3:
                        if (!(response.status === 400)) return [3 /*break*/, 5];
                        _b = {
                            status: 400
                        };
                        return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, (_b.value = _d.sent(),
                            _b)];
                    case 5:
                        if (!(response.status === 500)) return [3 /*break*/, 7];
                        _c = {
                            status: 500
                        };
                        return [4 /*yield*/, response.json()];
                    case 6: return [2 /*return*/, (_c.value = _d.sent(),
                            _c)];
                    case 7: throw response;
                }
            });
        });
    };
    /**
     */
    FlowNftOrderActivityControllerApi.prototype.getNftOrderAllActivities = function (requestParameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getNftOrderAllActivitiesRaw(requestParameters)];
                    case 1:
                        response = _a.sent();
                        if (response.status === 200) {
                            return [2 /*return*/, response.value];
                        }
                        throw response;
                }
            });
        });
    };
    return FlowNftOrderActivityControllerApi;
}(runtime.BaseAPI));
exports.FlowNftOrderActivityControllerApi = FlowNftOrderActivityControllerApi;
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByCollectionTypeEnum;
(function (GetNftOrderActivitiesByCollectionTypeEnum) {
    GetNftOrderActivitiesByCollectionTypeEnum["TRANSFER"] = "TRANSFER";
    GetNftOrderActivitiesByCollectionTypeEnum["MINT"] = "MINT";
    GetNftOrderActivitiesByCollectionTypeEnum["BURN"] = "BURN";
    GetNftOrderActivitiesByCollectionTypeEnum["LIST"] = "LIST";
    GetNftOrderActivitiesByCollectionTypeEnum["SELL"] = "SELL";
    GetNftOrderActivitiesByCollectionTypeEnum["CANCELLIST"] = "CANCEL_LIST";
})(GetNftOrderActivitiesByCollectionTypeEnum = exports.GetNftOrderActivitiesByCollectionTypeEnum || (exports.GetNftOrderActivitiesByCollectionTypeEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByCollectionSortEnum;
(function (GetNftOrderActivitiesByCollectionSortEnum) {
    GetNftOrderActivitiesByCollectionSortEnum["LATESTFIRST"] = "LATEST_FIRST";
    GetNftOrderActivitiesByCollectionSortEnum["EARLIESTFIRST"] = "EARLIEST_FIRST";
})(GetNftOrderActivitiesByCollectionSortEnum = exports.GetNftOrderActivitiesByCollectionSortEnum || (exports.GetNftOrderActivitiesByCollectionSortEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByItemTypeEnum;
(function (GetNftOrderActivitiesByItemTypeEnum) {
    GetNftOrderActivitiesByItemTypeEnum["TRANSFER"] = "TRANSFER";
    GetNftOrderActivitiesByItemTypeEnum["MINT"] = "MINT";
    GetNftOrderActivitiesByItemTypeEnum["BURN"] = "BURN";
    GetNftOrderActivitiesByItemTypeEnum["LIST"] = "LIST";
    GetNftOrderActivitiesByItemTypeEnum["SELL"] = "SELL";
    GetNftOrderActivitiesByItemTypeEnum["CANCELLIST"] = "CANCEL_LIST";
})(GetNftOrderActivitiesByItemTypeEnum = exports.GetNftOrderActivitiesByItemTypeEnum || (exports.GetNftOrderActivitiesByItemTypeEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByItemSortEnum;
(function (GetNftOrderActivitiesByItemSortEnum) {
    GetNftOrderActivitiesByItemSortEnum["LATESTFIRST"] = "LATEST_FIRST";
    GetNftOrderActivitiesByItemSortEnum["EARLIESTFIRST"] = "EARLIEST_FIRST";
})(GetNftOrderActivitiesByItemSortEnum = exports.GetNftOrderActivitiesByItemSortEnum || (exports.GetNftOrderActivitiesByItemSortEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByUserTypeEnum;
(function (GetNftOrderActivitiesByUserTypeEnum) {
    GetNftOrderActivitiesByUserTypeEnum["TRANSFERFROM"] = "TRANSFER_FROM";
    GetNftOrderActivitiesByUserTypeEnum["TRANSFERTO"] = "TRANSFER_TO";
    GetNftOrderActivitiesByUserTypeEnum["MINT"] = "MINT";
    GetNftOrderActivitiesByUserTypeEnum["BURN"] = "BURN";
    GetNftOrderActivitiesByUserTypeEnum["LIST"] = "LIST";
    GetNftOrderActivitiesByUserTypeEnum["BUY"] = "BUY";
    GetNftOrderActivitiesByUserTypeEnum["SELL"] = "SELL";
    GetNftOrderActivitiesByUserTypeEnum["CANCELLIST"] = "CANCEL_LIST";
})(GetNftOrderActivitiesByUserTypeEnum = exports.GetNftOrderActivitiesByUserTypeEnum || (exports.GetNftOrderActivitiesByUserTypeEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderActivitiesByUserSortEnum;
(function (GetNftOrderActivitiesByUserSortEnum) {
    GetNftOrderActivitiesByUserSortEnum["LATESTFIRST"] = "LATEST_FIRST";
    GetNftOrderActivitiesByUserSortEnum["EARLIESTFIRST"] = "EARLIEST_FIRST";
})(GetNftOrderActivitiesByUserSortEnum = exports.GetNftOrderActivitiesByUserSortEnum || (exports.GetNftOrderActivitiesByUserSortEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderAllActivitiesTypeEnum;
(function (GetNftOrderAllActivitiesTypeEnum) {
    GetNftOrderAllActivitiesTypeEnum["TRANSFER"] = "TRANSFER";
    GetNftOrderAllActivitiesTypeEnum["MINT"] = "MINT";
    GetNftOrderAllActivitiesTypeEnum["BURN"] = "BURN";
    GetNftOrderAllActivitiesTypeEnum["LIST"] = "LIST";
    GetNftOrderAllActivitiesTypeEnum["SELL"] = "SELL";
    GetNftOrderAllActivitiesTypeEnum["CANCELLIST"] = "CANCEL_LIST";
})(GetNftOrderAllActivitiesTypeEnum = exports.GetNftOrderAllActivitiesTypeEnum || (exports.GetNftOrderAllActivitiesTypeEnum = {}));
/**
    * @export
    * @enum {string}
    */
var GetNftOrderAllActivitiesSortEnum;
(function (GetNftOrderAllActivitiesSortEnum) {
    GetNftOrderAllActivitiesSortEnum["LATESTFIRST"] = "LATEST_FIRST";
    GetNftOrderAllActivitiesSortEnum["EARLIESTFIRST"] = "EARLIEST_FIRST";
})(GetNftOrderAllActivitiesSortEnum = exports.GetNftOrderAllActivitiesSortEnum || (exports.GetNftOrderAllActivitiesSortEnum = {}));
