// @rarible/protocol-ethereum-sdk/build/order/bid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderBid = void 0;
var tslib_1 = require("tslib");
var bn_1 = require("@rarible/utils/build/bn");
var action_1 = require("@rarible/action");
var types_1 = require("@rarible/types");
var OrderBid = /** @class */ (function () {
    function OrderBid(upserter, checkAssetType, checkWalletChainId) {
        var _this = this;
        this.upserter = upserter;
        this.checkAssetType = checkAssetType;
        this.checkWalletChainId = checkWalletChainId;
        this.bid = action_1.Action
            .create({
            id: "approve",
            run: function (request) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var form, checked;
                return (0, tslib_1.__generator)(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (request.makeAssetType.assetClass !== "ERC20") {
                                throw new Error("Make asset type should be ERC-20, received=".concat(request.makeAssetType.assetClass));
                            }
                            return [4 /*yield*/, this.getBidForm(request)];
                        case 1:
                            form = _a.sent();
                            return [4 /*yield*/, this.upserter.checkLazyOrder(form)];
                        case 2:
                            checked = _a.sent();
                            return [4 /*yield*/, this.upserter.approve(checked, true)];
                        case 3:
                            _a.sent();
                            return [2 /*return*/, checked];
                    }
                });
            }); },
        })
            .thenStep({
            id: "sign",
            run: function (checked) { return _this.upserter.upsertRequest(checked); },
        })
            .before(function (input) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkWalletChainId()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, input];
                }
            });
        }); });
        this.update = action_1.Action
            .create({
            id: "approve",
            run: function (request) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var order, price, form, checked;
                return (0, tslib_1.__generator)(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.upserter.getOrder(request)];
                        case 1:
                            order = _a.sent();
                            if (order.type === "CRYPTO_PUNK") {
                                return [2 /*return*/, request];
                            }
                            if (order.make.assetType.assetClass !== "ERC20") {
                                throw new Error("Make asset type should be ERC-20, received=".concat(order.make.assetType.assetClass));
                            }
                            return [4 /*yield*/, this.upserter.getPrice(request, order.make.assetType)];
                        case 2:
                            price = _a.sent();
                            return [4 /*yield*/, this.prepareOrderUpdateForm(order, price)];
                        case 3:
                            form = _a.sent();
                            return [4 /*yield*/, this.upserter.checkLazyOrder(form)];
                        case 4:
                            checked = _a.sent();
                            return [4 /*yield*/, this.upserter.approve(checked, true)];
                        case 5:
                            _a.sent();
                            return [2 /*return*/, checked];
                    }
                });
            }); },
        })
            .thenStep({
            id: "sign",
            run: function (form) {
                if ("type" in form && (form.type === "RARIBLE_V1" || form.type === "RARIBLE_V2")) {
                    return _this.upserter.upsertRequest(form);
                }
                return _this.upserter.updateCryptoPunkOrder(form);
            },
        })
            .before(function (input) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkWalletChainId()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, input];
                }
            });
        }); });
    }
    OrderBid.prototype.getBidForm = function (request) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var form, price, _a;
            var _b, _c;
            return (0, tslib_1.__generator)(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.upserter.prepareOrderForm(request, false)];
                    case 1:
                        form = _d.sent();
                        return [4 /*yield*/, this.upserter.getPrice(request, request.makeAssetType)];
                    case 2:
                        price = _d.sent();
                        _a = [(0, tslib_1.__assign)({}, form)];
                        _b = { make: {
                                assetType: request.makeAssetType,
                                value: (0, types_1.toBigNumber)((0, bn_1.toBn)(price).multipliedBy(request.amount).toString()),
                            } };
                        _c = {};
                        return [4 /*yield*/, this.checkAssetType(request.takeAssetType)];
                    case 3: return [2 /*return*/, tslib_1.__assign.apply(void 0, _a.concat([(_b.take = (_c.assetType = _d.sent(),
                                _c.value = (0, types_1.toBigNumber)(request.amount.toString()),
                                _c), _b)]))];
                }
            });
        });
    };
    OrderBid.prototype.prepareOrderUpdateForm = function (order, price) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                if (order.type === "RARIBLE_V1" || order.type === "RARIBLE_V2") {
                    return [2 /*return*/, this.upserter.getOrderFormFromOrder(order, {
                            assetType: order.make.assetType,
                            value: (0, types_1.toBigNumber)((0, bn_1.toBn)(price).multipliedBy(order.take.value).toString()),
                        }, order.take)];
                }
                throw new Error("Unsupported order type: ".concat(order.type));
            });
        });
    };
    return OrderBid;
}());
exports.OrderBid = OrderBid;
