// tezos-api-client/build/models/ActivitySort.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivitySortToJSON = exports.ActivitySortFromJSONTyped = exports.ActivitySortFromJSON = exports.ActivitySort = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var ActivitySort;
(function (ActivitySort) {
    ActivitySort["LatestFirst"] = "LATEST_FIRST";
    ActivitySort["EarliestFirst"] = "EARLIEST_FIRST";
})(ActivitySort = exports.ActivitySort || (exports.ActivitySort = {}));
function ActivitySortFromJSON(json) {
    return ActivitySortFromJSONTyped(json, false);
}
exports.ActivitySortFromJSON = ActivitySortFromJSON;
function ActivitySortFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.ActivitySortFromJSONTyped = ActivitySortFromJSONTyped;
function ActivitySortToJSON(value) {
    return value;
}
exports.ActivitySortToJSON = ActivitySortToJSON;
