// tezos-api-client/build/models/BadRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BadRequestToJSON = exports.BadRequestFromJSONTyped = exports.BadRequestFromJSON = exports.BadRequestCodeEnum = void 0;
/**
* @export
* @enum {string}
*/
var BadRequestCodeEnum;
(function (BadRequestCodeEnum) {
    BadRequestCodeEnum["BadRequest"] = "BAD_REQUEST";
    BadRequestCodeEnum["Validation"] = "VALIDATION";
})(BadRequestCodeEnum = exports.BadRequestCodeEnum || (exports.BadRequestCodeEnum = {}));
function BadRequestFromJSON(json) {
    return BadRequestFromJSONTyped(json, false);
}
exports.BadRequestFromJSON = BadRequestFromJSON;
function BadRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': json['code'],
        'message': json['message'],
    };
}
exports.BadRequestFromJSONTyped = BadRequestFromJSONTyped;
function BadRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
    };
}
exports.BadRequestToJSON = BadRequestToJSON;
