// tezos-api-client/build/models/Burn.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BurnToJSON = exports.BurnFromJSONTyped = exports.BurnFromJSON = exports.BurnTypeEnum = void 0;
/**
* @export
* @enum {string}
*/
var BurnTypeEnum;
(function (BurnTypeEnum) {
    BurnTypeEnum["Burn"] = "burn";
})(BurnTypeEnum = exports.BurnTypeEnum || (exports.BurnTypeEnum = {}));
function BurnFromJSON(json) {
    return BurnFromJSONTyped(json, false);
}
exports.BurnFromJSON = BurnFromJSON;
function BurnFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'owner': json['owner'],
        'contract': json['contract'],
        'tokenId': json['tokenId'],
        'value': json['value'],
        'transactionHash': json['transactionHash'],
        'blockHash': json['blockHash'],
        'blockNumber': json['blockNumber'],
    };
}
exports.BurnFromJSONTyped = BurnFromJSONTyped;
function BurnToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'owner': value.owner,
        'contract': value.contract,
        'tokenId': value.tokenId,
        'value': value.value,
        'transactionHash': value.transactionHash,
        'blockHash': value.blockHash,
        'blockNumber': value.blockNumber,
    };
}
exports.BurnToJSON = BurnToJSON;
