// tezos-api-client/build/models/EntityNotFound.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityNotFoundToJSON = exports.EntityNotFoundFromJSONTyped = exports.EntityNotFoundFromJSON = exports.EntityNotFoundCodeEnum = void 0;
/**
* @export
* @enum {string}
*/
var EntityNotFoundCodeEnum;
(function (EntityNotFoundCodeEnum) {
    EntityNotFoundCodeEnum["TokenNotFound"] = "TOKEN_NOT_FOUND";
    EntityNotFoundCodeEnum["BalanceNotFound"] = "BALANCE_NOT_FOUND";
    EntityNotFoundCodeEnum["OwnershipNotFound"] = "OWNERSHIP_NOT_FOUND";
    EntityNotFoundCodeEnum["OrderNotFound"] = "ORDER_NOT_FOUND";
    EntityNotFoundCodeEnum["ItemNotFound"] = "ITEM_NOT_FOUND";
    EntityNotFoundCodeEnum["CollectionNotFound"] = "COLLECTION_NOT_FOUND";
    EntityNotFoundCodeEnum["NotFound"] = "NOT_FOUND";
})(EntityNotFoundCodeEnum = exports.EntityNotFoundCodeEnum || (exports.EntityNotFoundCodeEnum = {}));
function EntityNotFoundFromJSON(json) {
    return EntityNotFoundFromJSONTyped(json, false);
}
exports.EntityNotFoundFromJSON = EntityNotFoundFromJSON;
function EntityNotFoundFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': json['code'],
        'message': json['message'],
    };
}
exports.EntityNotFoundFromJSONTyped = EntityNotFoundFromJSONTyped;
function EntityNotFoundToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
    };
}
exports.EntityNotFoundToJSON = EntityNotFoundToJSON;
