// tezos-api-client/build/models/FTAssetType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FTAssetTypeToJSON = exports.FTAssetTypeFromJSONTyped = exports.FTAssetTypeFromJSON = exports.FTAssetTypeAssetClassEnum = void 0;
var runtime_1 = require("../runtime");
/**
* @export
* @enum {string}
*/
var FTAssetTypeAssetClassEnum;
(function (FTAssetTypeAssetClassEnum) {
    FTAssetTypeAssetClassEnum["Ft"] = "FT";
})(FTAssetTypeAssetClassEnum = exports.FTAssetTypeAssetClassEnum || (exports.FTAssetTypeAssetClassEnum = {}));
function FTAssetTypeFromJSON(json) {
    return FTAssetTypeFromJSONTyped(json, false);
}
exports.FTAssetTypeFromJSON = FTAssetTypeFromJSON;
function FTAssetTypeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'assetClass': json['assetClass'],
        'contract': json['contract'],
        'tokenId': !(0, runtime_1.exists)(json, 'tokenId') ? undefined : json['tokenId'],
    };
}
exports.FTAssetTypeFromJSONTyped = FTAssetTypeFromJSONTyped;
function FTAssetTypeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'assetClass': value.assetClass,
        'contract': value.contract,
        'tokenId': value.tokenId,
    };
}
exports.FTAssetTypeToJSON = FTAssetTypeToJSON;
