// tezos-api-client/build/models/Mint.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintToJSON = exports.MintFromJSONTyped = exports.MintFromJSON = exports.MintTypeEnum = void 0;
/**
* @export
* @enum {string}
*/
var MintTypeEnum;
(function (MintTypeEnum) {
    MintTypeEnum["Mint"] = "mint";
})(MintTypeEnum = exports.MintTypeEnum || (exports.MintTypeEnum = {}));
function MintFromJSON(json) {
    return MintFromJSONTyped(json, false);
}
exports.MintFromJSON = MintFromJSON;
function MintFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'owner': json['owner'],
        'contract': json['contract'],
        'tokenId': json['tokenId'],
        'value': json['value'],
        'transactionHash': json['transactionHash'],
        'blockHash': json['blockHash'],
        'blockNumber': json['blockNumber'],
    };
}
exports.MintFromJSONTyped = MintFromJSONTyped;
function MintToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'owner': value.owner,
        'contract': value.contract,
        'tokenId': value.tokenId,
        'value': value.value,
        'transactionHash': value.transactionHash,
        'blockHash': value.blockHash,
        'blockNumber': value.blockNumber,
    };
}
exports.MintToJSON = MintToJSON;
