// tezos-api-client/build/models/MTAssetType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MTAssetTypeToJSON = exports.MTAssetTypeFromJSONTyped = exports.MTAssetTypeFromJSON = exports.MTAssetTypeAssetClassEnum = void 0;
/**
* @export
* @enum {string}
*/
var MTAssetTypeAssetClassEnum;
(function (MTAssetTypeAssetClassEnum) {
    MTAssetTypeAssetClassEnum["Mt"] = "MT";
})(MTAssetTypeAssetClassEnum = exports.MTAssetTypeAssetClassEnum || (exports.MTAssetTypeAssetClassEnum = {}));
function MTAssetTypeFromJSON(json) {
    return MTAssetTypeFromJSONTyped(json, false);
}
exports.MTAssetTypeFromJSON = MTAssetTypeFromJSON;
function MTAssetTypeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'assetClass': json['assetClass'],
        'contract': json['contract'],
        'tokenId': json['tokenId'],
    };
}
exports.MTAssetTypeFromJSONTyped = MTAssetTypeFromJSONTyped;
function MTAssetTypeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'assetClass': value.assetClass,
        'contract': value.contract,
        'tokenId': value.tokenId,
    };
}
exports.MTAssetTypeToJSON = MTAssetTypeToJSON;
