// tezos-api-client/build/models/NftActivityFilterByUser.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterByUserToJSON = exports.NftActivityFilterByUserFromJSONTyped = exports.NftActivityFilterByUserFromJSON = exports.NftActivityFilterByUserTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftActivityFilterByUserTypeEnum;
(function (NftActivityFilterByUserTypeEnum) {
    NftActivityFilterByUserTypeEnum["ByUser"] = "by_user";
})(NftActivityFilterByUserTypeEnum = exports.NftActivityFilterByUserTypeEnum || (exports.NftActivityFilterByUserTypeEnum = {}));
function NftActivityFilterByUserFromJSON(json) {
    return NftActivityFilterByUserFromJSONTyped(json, false);
}
exports.NftActivityFilterByUserFromJSON = NftActivityFilterByUserFromJSON;
function NftActivityFilterByUserFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.NftActivityFilterUserTypeFromJSON)),
        'users': json['users'],
    };
}
exports.NftActivityFilterByUserFromJSONTyped = NftActivityFilterByUserFromJSONTyped;
function NftActivityFilterByUserToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.NftActivityFilterUserTypeToJSON)),
        'users': value.users,
    };
}
exports.NftActivityFilterByUserToJSON = NftActivityFilterByUserToJSON;
