// tezos-api-client/build/models/NFTAssetType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NFTAssetTypeToJSON = exports.NFTAssetTypeFromJSONTyped = exports.NFTAssetTypeFromJSON = exports.NFTAssetTypeAssetClassEnum = void 0;
/**
* @export
* @enum {string}
*/
var NFTAssetTypeAssetClassEnum;
(function (NFTAssetTypeAssetClassEnum) {
    NFTAssetTypeAssetClassEnum["Nft"] = "NFT";
})(NFTAssetTypeAssetClassEnum = exports.NFTAssetTypeAssetClassEnum || (exports.NFTAssetTypeAssetClassEnum = {}));
function NFTAssetTypeFromJSON(json) {
    return NFTAssetTypeFromJSONTyped(json, false);
}
exports.NFTAssetTypeFromJSON = NFTAssetTypeFromJSON;
function NFTAssetTypeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'assetClass': json['assetClass'],
        'contract': json['contract'],
        'tokenId': json['tokenId'],
    };
}
exports.NFTAssetTypeFromJSONTyped = NFTAssetTypeFromJSONTyped;
function NFTAssetTypeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'assetClass': value.assetClass,
        'contract': value.contract,
        'tokenId': value.tokenId,
    };
}
exports.NFTAssetTypeToJSON = NFTAssetTypeToJSON;
