// tezos-api-client/build/models/NftActivityFilterAll.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterAllToJSON = exports.NftActivityFilterAllFromJSONTyped = exports.NftActivityFilterAllFromJSON = exports.NftActivityFilterAllTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftActivityFilterAllTypeEnum;
(function (NftActivityFilterAllTypeEnum) {
    NftActivityFilterAllTypeEnum["All"] = "all";
})(NftActivityFilterAllTypeEnum = exports.NftActivityFilterAllTypeEnum || (exports.NftActivityFilterAllTypeEnum = {}));
function NftActivityFilterAllFromJSON(json) {
    return NftActivityFilterAllFromJSONTyped(json, false);
}
exports.NftActivityFilterAllFromJSON = NftActivityFilterAllFromJSON;
function NftActivityFilterAllFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.NftActivityFilterAllTypeFromJSON)),
    };
}
exports.NftActivityFilterAllFromJSONTyped = NftActivityFilterAllFromJSONTyped;
function NftActivityFilterAllToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.NftActivityFilterAllTypeToJSON)),
    };
}
exports.NftActivityFilterAllToJSON = NftActivityFilterAllToJSON;
