// tezos-api-client/build/models/NftActivityFilterAllType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterAllTypeToJSON = exports.NftActivityFilterAllTypeFromJSONTyped = exports.NftActivityFilterAllTypeFromJSON = exports.NftActivityFilterAllType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var NftActivityFilterAllType;
(function (NftActivityFilterAllType) {
    NftActivityFilterAllType["Transfer"] = "TRANSFER";
    NftActivityFilterAllType["Mint"] = "MINT";
    NftActivityFilterAllType["Burn"] = "BURN";
})(NftActivityFilterAllType = exports.NftActivityFilterAllType || (exports.NftActivityFilterAllType = {}));
function NftActivityFilterAllTypeFromJSON(json) {
    return NftActivityFilterAllTypeFromJSONTyped(json, false);
}
exports.NftActivityFilterAllTypeFromJSON = NftActivityFilterAllTypeFromJSON;
function NftActivityFilterAllTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.NftActivityFilterAllTypeFromJSONTyped = NftActivityFilterAllTypeFromJSONTyped;
function NftActivityFilterAllTypeToJSON(value) {
    return value;
}
exports.NftActivityFilterAllTypeToJSON = NftActivityFilterAllTypeToJSON;
