// tezos-api-client/build/models/NftActivityFilterByItem.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterByItemToJSON = exports.NftActivityFilterByItemFromJSONTyped = exports.NftActivityFilterByItemFromJSON = exports.NftActivityFilterByItemTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftActivityFilterByItemTypeEnum;
(function (NftActivityFilterByItemTypeEnum) {
    NftActivityFilterByItemTypeEnum["ByItem"] = "by_item";
})(NftActivityFilterByItemTypeEnum = exports.NftActivityFilterByItemTypeEnum || (exports.NftActivityFilterByItemTypeEnum = {}));
function NftActivityFilterByItemFromJSON(json) {
    return NftActivityFilterByItemFromJSONTyped(json, false);
}
exports.NftActivityFilterByItemFromJSON = NftActivityFilterByItemFromJSON;
function NftActivityFilterByItemFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.NftActivityFilterAllTypeFromJSON)),
        'contract': json['contract'],
        'tokenId': json['tokenId'],
    };
}
exports.NftActivityFilterByItemFromJSONTyped = NftActivityFilterByItemFromJSONTyped;
function NftActivityFilterByItemToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.NftActivityFilterAllTypeToJSON)),
        'contract': value.contract,
        'tokenId': value.tokenId,
    };
}
exports.NftActivityFilterByItemToJSON = NftActivityFilterByItemToJSON;
