// tezos-api-client/build/models/NftActivityFilterUserType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftActivityFilterUserTypeToJSON = exports.NftActivityFilterUserTypeFromJSONTyped = exports.NftActivityFilterUserTypeFromJSON = exports.NftActivityFilterUserType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var NftActivityFilterUserType;
(function (NftActivityFilterUserType) {
    NftActivityFilterUserType["TransferFrom"] = "TRANSFER_FROM";
    NftActivityFilterUserType["TransferTo"] = "TRANSFER_TO";
    NftActivityFilterUserType["Mint"] = "MINT";
    NftActivityFilterUserType["Burn"] = "BURN";
})(NftActivityFilterUserType = exports.NftActivityFilterUserType || (exports.NftActivityFilterUserType = {}));
function NftActivityFilterUserTypeFromJSON(json) {
    return NftActivityFilterUserTypeFromJSONTyped(json, false);
}
exports.NftActivityFilterUserTypeFromJSON = NftActivityFilterUserTypeFromJSON;
function NftActivityFilterUserTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.NftActivityFilterUserTypeFromJSONTyped = NftActivityFilterUserTypeFromJSONTyped;
function NftActivityFilterUserTypeToJSON(value) {
    return value;
}
exports.NftActivityFilterUserTypeToJSON = NftActivityFilterUserTypeToJSON;
