// tezos-api-client/build/models/NftCollectionType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftCollectionTypeToJSON = exports.NftCollectionTypeFromJSONTyped = exports.NftCollectionTypeFromJSON = exports.NftCollectionType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var NftCollectionType;
(function (NftCollectionType) {
    NftCollectionType["Nft"] = "NFT";
    NftCollectionType["Mt"] = "MT";
})(NftCollectionType = exports.NftCollectionType || (exports.NftCollectionType = {}));
function NftCollectionTypeFromJSON(json) {
    return NftCollectionTypeFromJSONTyped(json, false);
}
exports.NftCollectionTypeFromJSON = NftCollectionTypeFromJSON;
function NftCollectionTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.NftCollectionTypeFromJSONTyped = NftCollectionTypeFromJSONTyped;
function NftCollectionTypeToJSON(value) {
    return value;
}
exports.NftCollectionTypeToJSON = NftCollectionTypeToJSON;
