// tezos-api-client/build/models/NftCollectionEvent.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftCollectionEventToJSON = exports.NftCollectionEventFromJSONTyped = exports.NftCollectionEventFromJSON = exports.NftCollectionEventTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftCollectionEventTypeEnum;
(function (NftCollectionEventTypeEnum) {
    NftCollectionEventTypeEnum["Update"] = "UPDATE";
})(NftCollectionEventTypeEnum = exports.NftCollectionEventTypeEnum || (exports.NftCollectionEventTypeEnum = {}));
function NftCollectionEventFromJSON(json) {
    return NftCollectionEventFromJSONTyped(json, false);
}
exports.NftCollectionEventFromJSON = NftCollectionEventFromJSON;
function NftCollectionEventFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'eventId': json['eventId'],
        'collectionId': json['collectionId'],
        'type': json['type'],
        'collection': (0, _1.NftCollectionFromJSON)(json['collection']),
    };
}
exports.NftCollectionEventFromJSONTyped = NftCollectionEventFromJSONTyped;
function NftCollectionEventToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'eventId': value.eventId,
        'collectionId': value.collectionId,
        'type': value.type,
        'collection': (0, _1.NftCollectionToJSON)(value.collection),
    };
}
exports.NftCollectionEventToJSON = NftCollectionEventToJSON;
