// tezos-api-client/build/models/NftCollectionFeature.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftCollectionFeatureToJSON = exports.NftCollectionFeatureFromJSONTyped = exports.NftCollectionFeatureFromJSON = exports.NftCollectionFeature = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var NftCollectionFeature;
(function (NftCollectionFeature) {
    NftCollectionFeature["ApproveForAll"] = "APPROVE_FOR_ALL";
    NftCollectionFeature["SetUriPrefix"] = "SET_URI_PREFIX";
    NftCollectionFeature["Burn"] = "BURN";
    NftCollectionFeature["MintWithAddress"] = "MINT_WITH_ADDRESS";
    NftCollectionFeature["SecondarySaleFees"] = "SECONDARY_SALE_FEES";
    NftCollectionFeature["MintAndTransfer"] = "MINT_AND_TRANSFER";
})(NftCollectionFeature = exports.NftCollectionFeature || (exports.NftCollectionFeature = {}));
function NftCollectionFeatureFromJSON(json) {
    return NftCollectionFeatureFromJSONTyped(json, false);
}
exports.NftCollectionFeatureFromJSON = NftCollectionFeatureFromJSON;
function NftCollectionFeatureFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.NftCollectionFeatureFromJSONTyped = NftCollectionFeatureFromJSONTyped;
function NftCollectionFeatureToJSON(value) {
    return value;
}
exports.NftCollectionFeatureToJSON = NftCollectionFeatureToJSON;
