// tezos-api-client/build/models/NftItemEvent.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftItemEventToJSON = exports.NftItemEventFromJSONTyped = exports.NftItemEventFromJSON = exports.NftItemEventTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftItemEventTypeEnum;
(function (NftItemEventTypeEnum) {
    NftItemEventTypeEnum["Update"] = "UPDATE";
    NftItemEventTypeEnum["Delete"] = "DELETE";
})(NftItemEventTypeEnum = exports.NftItemEventTypeEnum || (exports.NftItemEventTypeEnum = {}));
function NftItemEventFromJSON(json) {
    return NftItemEventFromJSONTyped(json, false);
}
exports.NftItemEventFromJSON = NftItemEventFromJSON;
function NftItemEventFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'eventId': json['eventId'],
        'itemId': json['itemId'],
        'type': json['type'],
        'item': (0, _1.NftItemFromJSON)(json['item']),
    };
}
exports.NftItemEventFromJSONTyped = NftItemEventFromJSONTyped;
function NftItemEventToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'eventId': value.eventId,
        'itemId': value.itemId,
        'type': value.type,
        'item': (0, _1.NftItemToJSON)(value.item),
    };
}
exports.NftItemEventToJSON = NftItemEventToJSON;
