// tezos-api-client/build/models/NftOwnershipEvent.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftOwnershipEventToJSON = exports.NftOwnershipEventFromJSONTyped = exports.NftOwnershipEventFromJSON = exports.NftOwnershipEventTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var NftOwnershipEventTypeEnum;
(function (NftOwnershipEventTypeEnum) {
    NftOwnershipEventTypeEnum["Update"] = "UPDATE";
    NftOwnershipEventTypeEnum["Delete"] = "DELETE";
})(NftOwnershipEventTypeEnum = exports.NftOwnershipEventTypeEnum || (exports.NftOwnershipEventTypeEnum = {}));
function NftOwnershipEventFromJSON(json) {
    return NftOwnershipEventFromJSONTyped(json, false);
}
exports.NftOwnershipEventFromJSON = NftOwnershipEventFromJSON;
function NftOwnershipEventFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'eventId': json['eventId'],
        'ownershipId': json['ownershipId'],
        'type': json['type'],
        'ownership': (0, _1.NftOwnershipFromJSON)(json['ownership']),
    };
}
exports.NftOwnershipEventFromJSONTyped = NftOwnershipEventFromJSONTyped;
function NftOwnershipEventToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'eventId': value.eventId,
        'ownershipId': value.ownershipId,
        'type': value.type,
        'ownership': (0, _1.NftOwnershipToJSON)(value.ownership),
    };
}
exports.NftOwnershipEventToJSON = NftOwnershipEventToJSON;
