// tezos-api-client/build/models/Order.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderToJSON = exports.OrderFromJSONTyped = exports.OrderFromJSON = exports.OrderTypeEnum = void 0;
var runtime_1 = require("../runtime");
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderTypeEnum;
(function (OrderTypeEnum) {
    OrderTypeEnum["RaribleV2"] = "RARIBLE_V2";
})(OrderTypeEnum = exports.OrderTypeEnum || (exports.OrderTypeEnum = {}));
function OrderFromJSON(json) {
    return OrderFromJSONTyped(json, false);
}
exports.OrderFromJSON = OrderFromJSON;
function OrderFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'data': (0, _1.OrderRaribleV2DataV1FromJSON)(json['data']),
        'maker': json['maker'],
        'makerEdpk': json['makerEdpk'],
        'taker': !(0, runtime_1.exists)(json, 'taker') ? undefined : json['taker'],
        'takerEdpk': !(0, runtime_1.exists)(json, 'takerEdpk') ? undefined : json['takerEdpk'],
        'make': (0, _1.AssetFromJSON)(json['make']),
        'take': (0, _1.AssetFromJSON)(json['take']),
        'fill': json['fill'],
        'start': !(0, runtime_1.exists)(json, 'start') ? undefined : json['start'],
        'end': !(0, runtime_1.exists)(json, 'end') ? undefined : json['end'],
        'makeStock': json['makeStock'],
        'cancelled': json['cancelled'],
        'salt': json['salt'],
        'signature': json['signature'],
        'createdAt': (new Date(json['createdAt'])),
        'lastUpdateAt': (new Date(json['lastUpdateAt'])),
        'hash': json['hash'],
        'makeBalance': !(0, runtime_1.exists)(json, 'makeBalance') ? undefined : json['makeBalance'],
        'priceHistory': !(0, runtime_1.exists)(json, 'priceHistory') ? undefined : (json['priceHistory'].map(_1.OrderPriceHistoryRecordFromJSON)),
        'status': (0, _1.OrderStatusFromJSON)(json['status']),
        'type': json['type'],
    };
}
exports.OrderFromJSONTyped = OrderFromJSONTyped;
function OrderToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'data': (0, _1.OrderRaribleV2DataV1ToJSON)(value.data),
        'maker': value.maker,
        'makerEdpk': value.makerEdpk,
        'taker': value.taker,
        'takerEdpk': value.takerEdpk,
        'make': (0, _1.AssetToJSON)(value.make),
        'take': (0, _1.AssetToJSON)(value.take),
        'fill': value.fill,
        'start': value.start,
        'end': value.end,
        'makeStock': value.makeStock,
        'cancelled': value.cancelled,
        'salt': value.salt,
        'signature': value.signature,
        'createdAt': (value.createdAt.toISOString()),
        'lastUpdateAt': (value.lastUpdateAt.toISOString()),
        'hash': value.hash,
        'makeBalance': value.makeBalance,
        'priceHistory': value.priceHistory === undefined ? undefined : (value.priceHistory.map(_1.OrderPriceHistoryRecordToJSON)),
        'status': (0, _1.OrderStatusToJSON)(value.status),
        'type': value.type,
    };
}
exports.OrderToJSON = OrderToJSON;
