// tezos-api-client/build/models/OrderActivityMatch.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityMatchToJSON = exports.OrderActivityMatchFromJSONTyped = exports.OrderActivityMatchFromJSON = exports.OrderActivityMatchTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityMatchTypeEnum;
(function (OrderActivityMatchTypeEnum) {
    OrderActivityMatchTypeEnum["Match"] = "match";
})(OrderActivityMatchTypeEnum = exports.OrderActivityMatchTypeEnum || (exports.OrderActivityMatchTypeEnum = {}));
function OrderActivityMatchFromJSON(json) {
    return OrderActivityMatchFromJSONTyped(json, false);
}
exports.OrderActivityMatchFromJSON = OrderActivityMatchFromJSON;
function OrderActivityMatchFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'left': (0, _1.OrderActivitySideMatchFromJSON)(json['left']),
        'right': (0, _1.OrderActivitySideMatchFromJSON)(json['right']),
        'type': (0, _1.OrderActivityMatchTypeFromJSON)(json['type']),
        'price': json['price'],
        'transactionHash': json['transactionHash'],
        'blockHash': json['blockHash'],
        'blockNumber': json['blockNumber'],
        'logIndex': json['logIndex'],
    };
}
exports.OrderActivityMatchFromJSONTyped = OrderActivityMatchFromJSONTyped;
function OrderActivityMatchToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'left': (0, _1.OrderActivitySideMatchToJSON)(value.left),
        'right': (0, _1.OrderActivitySideMatchToJSON)(value.right),
        'type': (0, _1.OrderActivityMatchTypeToJSON)(value.type),
        'price': value.price,
        'transactionHash': value.transactionHash,
        'blockHash': value.blockHash,
        'blockNumber': value.blockNumber,
        'logIndex': value.logIndex,
    };
}
exports.OrderActivityMatchToJSON = OrderActivityMatchToJSON;
