// tezos-api-client/build/models/OrderActivityFilterByItem.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterByItemToJSON = exports.OrderActivityFilterByItemFromJSONTyped = exports.OrderActivityFilterByItemFromJSON = exports.OrderActivityFilterByItemTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityFilterByItemTypeEnum;
(function (OrderActivityFilterByItemTypeEnum) {
    OrderActivityFilterByItemTypeEnum["ByItem"] = "by_item";
})(OrderActivityFilterByItemTypeEnum = exports.OrderActivityFilterByItemTypeEnum || (exports.OrderActivityFilterByItemTypeEnum = {}));
function OrderActivityFilterByItemFromJSON(json) {
    return OrderActivityFilterByItemFromJSONTyped(json, false);
}
exports.OrderActivityFilterByItemFromJSON = OrderActivityFilterByItemFromJSON;
function OrderActivityFilterByItemFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.OrderActivityFilterAllTypeFromJSON)),
        'contract': json['contract'],
        'tokenId': json['tokenId'],
    };
}
exports.OrderActivityFilterByItemFromJSONTyped = OrderActivityFilterByItemFromJSONTyped;
function OrderActivityFilterByItemToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.OrderActivityFilterAllTypeToJSON)),
        'contract': value.contract,
        'tokenId': value.tokenId,
    };
}
exports.OrderActivityFilterByItemToJSON = OrderActivityFilterByItemToJSON;
