// tezos-api-client/build/models/OrderActivityCancelBid.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityCancelBidToJSON = exports.OrderActivityCancelBidFromJSONTyped = exports.OrderActivityCancelBidFromJSON = exports.OrderActivityCancelBidTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityCancelBidTypeEnum;
(function (OrderActivityCancelBidTypeEnum) {
    OrderActivityCancelBidTypeEnum["CancelBid"] = "cancel_bid";
})(OrderActivityCancelBidTypeEnum = exports.OrderActivityCancelBidTypeEnum || (exports.OrderActivityCancelBidTypeEnum = {}));
function OrderActivityCancelBidFromJSON(json) {
    return OrderActivityCancelBidFromJSONTyped(json, false);
}
exports.OrderActivityCancelBidFromJSON = OrderActivityCancelBidFromJSON;
function OrderActivityCancelBidFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'hash': json['hash'],
        'maker': json['maker'],
        'make': (0, _1.AssetTypeFromJSON)(json['make']),
        'take': (0, _1.AssetTypeFromJSON)(json['take']),
        'transactionHash': json['transactionHash'],
        'blockHash': json['blockHash'],
        'blockNumber': json['blockNumber'],
        'logIndex': json['logIndex'],
    };
}
exports.OrderActivityCancelBidFromJSONTyped = OrderActivityCancelBidFromJSONTyped;
function OrderActivityCancelBidToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'hash': value.hash,
        'maker': value.maker,
        'make': (0, _1.AssetTypeToJSON)(value.make),
        'take': (0, _1.AssetTypeToJSON)(value.take),
        'transactionHash': value.transactionHash,
        'blockHash': value.blockHash,
        'blockNumber': value.blockNumber,
        'logIndex': value.logIndex,
    };
}
exports.OrderActivityCancelBidToJSON = OrderActivityCancelBidToJSON;
