// tezos-api-client/build/models/OrderForm.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderFormToJSON = exports.OrderFormFromJSONTyped = exports.OrderFormFromJSON = exports.OrderFormTypeEnum = void 0;
var runtime_1 = require("../runtime");
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderFormTypeEnum;
(function (OrderFormTypeEnum) {
    OrderFormTypeEnum["RaribleV2"] = "RARIBLE_V2";
})(OrderFormTypeEnum = exports.OrderFormTypeEnum || (exports.OrderFormTypeEnum = {}));
function OrderFormFromJSON(json) {
    return OrderFormFromJSONTyped(json, false);
}
exports.OrderFormFromJSON = OrderFormFromJSON;
function OrderFormFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'data': (0, _1.OrderRaribleV2DataV1FromJSON)(json['data']),
        'maker': json['maker'],
        'makerEdpk': json['makerEdpk'],
        'taker': !(0, runtime_1.exists)(json, 'taker') ? undefined : json['taker'],
        'takerEdpk': !(0, runtime_1.exists)(json, 'takerEdpk') ? undefined : json['takerEdpk'],
        'make': (0, _1.AssetFromJSON)(json['make']),
        'take': (0, _1.AssetFromJSON)(json['take']),
        'salt': json['salt'],
        'start': !(0, runtime_1.exists)(json, 'start') ? undefined : json['start'],
        'end': !(0, runtime_1.exists)(json, 'end') ? undefined : json['end'],
        'signature': json['signature'],
        'type': json['type'],
    };
}
exports.OrderFormFromJSONTyped = OrderFormFromJSONTyped;
function OrderFormToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'data': (0, _1.OrderRaribleV2DataV1ToJSON)(value.data),
        'maker': value.maker,
        'makerEdpk': value.makerEdpk,
        'taker': value.taker,
        'takerEdpk': value.takerEdpk,
        'make': (0, _1.AssetToJSON)(value.make),
        'take': (0, _1.AssetToJSON)(value.take),
        'salt': value.salt,
        'start': value.start,
        'end': value.end,
        'signature': value.signature,
        'type': value.type,
    };
}
exports.OrderFormToJSON = OrderFormToJSON;
