// tezos-api-client/build/models/OrderActivityFilterByUser.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterByUserToJSON = exports.OrderActivityFilterByUserFromJSONTyped = exports.OrderActivityFilterByUserFromJSON = exports.OrderActivityFilterByUserTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityFilterByUserTypeEnum;
(function (OrderActivityFilterByUserTypeEnum) {
    OrderActivityFilterByUserTypeEnum["ByUser"] = "by_user";
})(OrderActivityFilterByUserTypeEnum = exports.OrderActivityFilterByUserTypeEnum || (exports.OrderActivityFilterByUserTypeEnum = {}));
function OrderActivityFilterByUserFromJSON(json) {
    return OrderActivityFilterByUserFromJSONTyped(json, false);
}
exports.OrderActivityFilterByUserFromJSON = OrderActivityFilterByUserFromJSON;
function OrderActivityFilterByUserFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.OrderActivityFilterUserTypeFromJSON)),
        'users': json['users'],
    };
}
exports.OrderActivityFilterByUserFromJSONTyped = OrderActivityFilterByUserFromJSONTyped;
function OrderActivityFilterByUserToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.OrderActivityFilterUserTypeToJSON)),
        'users': value.users,
    };
}
exports.OrderActivityFilterByUserToJSON = OrderActivityFilterByUserToJSON;
