// tezos-api-client/build/models/OrderActivityFilterAllType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterAllTypeToJSON = exports.OrderActivityFilterAllTypeFromJSONTyped = exports.OrderActivityFilterAllTypeFromJSON = exports.OrderActivityFilterAllType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var OrderActivityFilterAllType;
(function (OrderActivityFilterAllType) {
    OrderActivityFilterAllType["Bid"] = "BID";
    OrderActivityFilterAllType["List"] = "LIST";
    OrderActivityFilterAllType["Match"] = "MATCH";
    OrderActivityFilterAllType["CancelList"] = "CANCEL_LIST";
    OrderActivityFilterAllType["CancelBid"] = "CANCEL_BID";
})(OrderActivityFilterAllType = exports.OrderActivityFilterAllType || (exports.OrderActivityFilterAllType = {}));
function OrderActivityFilterAllTypeFromJSON(json) {
    return OrderActivityFilterAllTypeFromJSONTyped(json, false);
}
exports.OrderActivityFilterAllTypeFromJSON = OrderActivityFilterAllTypeFromJSON;
function OrderActivityFilterAllTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.OrderActivityFilterAllTypeFromJSONTyped = OrderActivityFilterAllTypeFromJSONTyped;
function OrderActivityFilterAllTypeToJSON(value) {
    return value;
}
exports.OrderActivityFilterAllTypeToJSON = OrderActivityFilterAllTypeToJSON;
