// tezos-api-client/build/models/OrderActivityCancelList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityCancelListToJSON = exports.OrderActivityCancelListFromJSONTyped = exports.OrderActivityCancelListFromJSON = exports.OrderActivityCancelListTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityCancelListTypeEnum;
(function (OrderActivityCancelListTypeEnum) {
    OrderActivityCancelListTypeEnum["CancelList"] = "cancel_list";
})(OrderActivityCancelListTypeEnum = exports.OrderActivityCancelListTypeEnum || (exports.OrderActivityCancelListTypeEnum = {}));
function OrderActivityCancelListFromJSON(json) {
    return OrderActivityCancelListFromJSONTyped(json, false);
}
exports.OrderActivityCancelListFromJSON = OrderActivityCancelListFromJSON;
function OrderActivityCancelListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'hash': json['hash'],
        'maker': json['maker'],
        'make': (0, _1.AssetTypeFromJSON)(json['make']),
        'take': (0, _1.AssetTypeFromJSON)(json['take']),
        'transactionHash': json['transactionHash'],
        'blockHash': json['blockHash'],
        'blockNumber': json['blockNumber'],
        'logIndex': json['logIndex'],
    };
}
exports.OrderActivityCancelListFromJSONTyped = OrderActivityCancelListFromJSONTyped;
function OrderActivityCancelListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'hash': value.hash,
        'maker': value.maker,
        'make': (0, _1.AssetTypeToJSON)(value.make),
        'take': (0, _1.AssetTypeToJSON)(value.take),
        'transactionHash': value.transactionHash,
        'blockHash': value.blockHash,
        'blockNumber': value.blockNumber,
        'logIndex': value.logIndex,
    };
}
exports.OrderActivityCancelListToJSON = OrderActivityCancelListToJSON;
