// tezos-api-client/build/models/OrderActivityFilterUserType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterUserTypeToJSON = exports.OrderActivityFilterUserTypeFromJSONTyped = exports.OrderActivityFilterUserTypeFromJSON = exports.OrderActivityFilterUserType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var OrderActivityFilterUserType;
(function (OrderActivityFilterUserType) {
    OrderActivityFilterUserType["CancelList"] = "CANCEL_LIST";
    OrderActivityFilterUserType["CancelBid"] = "CANCEL_BID";
    OrderActivityFilterUserType["MakeBid"] = "MAKE_BID";
    OrderActivityFilterUserType["GetBid"] = "GET_BID";
    OrderActivityFilterUserType["List"] = "LIST";
    OrderActivityFilterUserType["Buy"] = "BUY";
    OrderActivityFilterUserType["Sell"] = "SELL";
})(OrderActivityFilterUserType = exports.OrderActivityFilterUserType || (exports.OrderActivityFilterUserType = {}));
function OrderActivityFilterUserTypeFromJSON(json) {
    return OrderActivityFilterUserTypeFromJSONTyped(json, false);
}
exports.OrderActivityFilterUserTypeFromJSON = OrderActivityFilterUserTypeFromJSON;
function OrderActivityFilterUserTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.OrderActivityFilterUserTypeFromJSONTyped = OrderActivityFilterUserTypeFromJSONTyped;
function OrderActivityFilterUserTypeToJSON(value) {
    return value;
}
exports.OrderActivityFilterUserTypeToJSON = OrderActivityFilterUserTypeToJSON;
