// tezos-api-client/build/models/OrderType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderTypeToJSON = exports.OrderTypeFromJSONTyped = exports.OrderTypeFromJSON = exports.OrderType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var OrderType;
(function (OrderType) {
    OrderType["Sell"] = "SELL";
    OrderType["Bid"] = "BID";
})(OrderType = exports.OrderType || (exports.OrderType = {}));
function OrderTypeFromJSON(json) {
    return OrderTypeFromJSONTyped(json, false);
}
exports.OrderTypeFromJSON = OrderTypeFromJSON;
function OrderTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.OrderTypeFromJSONTyped = OrderTypeFromJSONTyped;
function OrderTypeToJSON(value) {
    return value;
}
exports.OrderTypeToJSON = OrderTypeToJSON;
