// tezos-api-client/build/models/OrderActivityFilterAll.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterAllToJSON = exports.OrderActivityFilterAllFromJSONTyped = exports.OrderActivityFilterAllFromJSON = exports.OrderActivityFilterAllTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityFilterAllTypeEnum;
(function (OrderActivityFilterAllTypeEnum) {
    OrderActivityFilterAllTypeEnum["All"] = "all";
})(OrderActivityFilterAllTypeEnum = exports.OrderActivityFilterAllTypeEnum || (exports.OrderActivityFilterAllTypeEnum = {}));
function OrderActivityFilterAllFromJSON(json) {
    return OrderActivityFilterAllFromJSONTyped(json, false);
}
exports.OrderActivityFilterAllFromJSON = OrderActivityFilterAllFromJSON;
function OrderActivityFilterAllFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.OrderActivityFilterAllTypeFromJSON)),
    };
}
exports.OrderActivityFilterAllFromJSONTyped = OrderActivityFilterAllFromJSONTyped;
function OrderActivityFilterAllToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.OrderActivityFilterAllTypeToJSON)),
    };
}
exports.OrderActivityFilterAllToJSON = OrderActivityFilterAllToJSON;
