// tezos-api-client/build/models/OrderEvent.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderEventToJSON = exports.OrderEventFromJSONTyped = exports.OrderEventFromJSON = exports.OrderEventTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderEventTypeEnum;
(function (OrderEventTypeEnum) {
    OrderEventTypeEnum["Update"] = "UPDATE";
})(OrderEventTypeEnum = exports.OrderEventTypeEnum || (exports.OrderEventTypeEnum = {}));
function OrderEventFromJSON(json) {
    return OrderEventFromJSONTyped(json, false);
}
exports.OrderEventFromJSON = OrderEventFromJSON;
function OrderEventFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'eventId': json['eventId'],
        'orderId': json['orderId'],
        'type': json['type'],
        'order': (0, _1.OrderFromJSON)(json['order']),
    };
}
exports.OrderEventFromJSONTyped = OrderEventFromJSONTyped;
function OrderEventToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'eventId': value.eventId,
        'orderId': value.orderId,
        'type': value.type,
        'order': (0, _1.OrderToJSON)(value.order),
    };
}
exports.OrderEventToJSON = OrderEventToJSON;
