// tezos-api-client/build/models/OrderActivitySideType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivitySideTypeToJSON = exports.OrderActivitySideTypeFromJSONTyped = exports.OrderActivitySideTypeFromJSON = exports.OrderActivitySideType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var OrderActivitySideType;
(function (OrderActivitySideType) {
    OrderActivitySideType["Sell"] = "SELL";
    OrderActivitySideType["Bid"] = "BID";
})(OrderActivitySideType = exports.OrderActivitySideType || (exports.OrderActivitySideType = {}));
function OrderActivitySideTypeFromJSON(json) {
    return OrderActivitySideTypeFromJSONTyped(json, false);
}
exports.OrderActivitySideTypeFromJSON = OrderActivitySideTypeFromJSON;
function OrderActivitySideTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.OrderActivitySideTypeFromJSONTyped = OrderActivitySideTypeFromJSONTyped;
function OrderActivitySideTypeToJSON(value) {
    return value;
}
exports.OrderActivitySideTypeToJSON = OrderActivitySideTypeToJSON;
