// tezos-api-client/build/models/OrderActivityMatchType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityMatchTypeToJSON = exports.OrderActivityMatchTypeFromJSONTyped = exports.OrderActivityMatchTypeFromJSON = exports.OrderActivityMatchType = void 0;
/**
 *
 * @export
 * @enum {string}
 */
var OrderActivityMatchType;
(function (OrderActivityMatchType) {
    OrderActivityMatchType["Sell"] = "SELL";
    OrderActivityMatchType["AcceptBid"] = "ACCEPT_BID";
})(OrderActivityMatchType = exports.OrderActivityMatchType || (exports.OrderActivityMatchType = {}));
function OrderActivityMatchTypeFromJSON(json) {
    return OrderActivityMatchTypeFromJSONTyped(json, false);
}
exports.OrderActivityMatchTypeFromJSON = OrderActivityMatchTypeFromJSON;
function OrderActivityMatchTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.OrderActivityMatchTypeFromJSONTyped = OrderActivityMatchTypeFromJSONTyped;
function OrderActivityMatchTypeToJSON(value) {
    return value;
}
exports.OrderActivityMatchTypeToJSON = OrderActivityMatchTypeToJSON;
