// tezos-api-client/build/models/OrderActivityFilterByCollection.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityFilterByCollectionToJSON = exports.OrderActivityFilterByCollectionFromJSONTyped = exports.OrderActivityFilterByCollectionFromJSON = exports.OrderActivityFilterByCollectionTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityFilterByCollectionTypeEnum;
(function (OrderActivityFilterByCollectionTypeEnum) {
    OrderActivityFilterByCollectionTypeEnum["ByCollection"] = "by_collection";
})(OrderActivityFilterByCollectionTypeEnum = exports.OrderActivityFilterByCollectionTypeEnum || (exports.OrderActivityFilterByCollectionTypeEnum = {}));
function OrderActivityFilterByCollectionFromJSON(json) {
    return OrderActivityFilterByCollectionFromJSONTyped(json, false);
}
exports.OrderActivityFilterByCollectionFromJSON = OrderActivityFilterByCollectionFromJSON;
function OrderActivityFilterByCollectionFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'types': (json['types'].map(_1.OrderActivityFilterAllTypeFromJSON)),
        'contract': json['contract'],
    };
}
exports.OrderActivityFilterByCollectionFromJSONTyped = OrderActivityFilterByCollectionFromJSONTyped;
function OrderActivityFilterByCollectionToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'types': (value.types.map(_1.OrderActivityFilterAllTypeToJSON)),
        'contract': value.contract,
    };
}
exports.OrderActivityFilterByCollectionToJSON = OrderActivityFilterByCollectionToJSON;
