// tezos-api-client/build/models/OrderActivityBid.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityBidToJSON = exports.OrderActivityBidFromJSONTyped = exports.OrderActivityBidFromJSON = exports.OrderActivityBidTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityBidTypeEnum;
(function (OrderActivityBidTypeEnum) {
    OrderActivityBidTypeEnum["Bid"] = "bid";
})(OrderActivityBidTypeEnum = exports.OrderActivityBidTypeEnum || (exports.OrderActivityBidTypeEnum = {}));
function OrderActivityBidFromJSON(json) {
    return OrderActivityBidFromJSONTyped(json, false);
}
exports.OrderActivityBidFromJSON = OrderActivityBidFromJSON;
function OrderActivityBidFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'hash': json['hash'],
        'maker': json['maker'],
        'make': (0, _1.AssetFromJSON)(json['make']),
        'take': (0, _1.AssetFromJSON)(json['take']),
        'price': json['price'],
    };
}
exports.OrderActivityBidFromJSONTyped = OrderActivityBidFromJSONTyped;
function OrderActivityBidToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'hash': value.hash,
        'maker': value.maker,
        'make': (0, _1.AssetToJSON)(value.make),
        'take': (0, _1.AssetToJSON)(value.take),
        'price': value.price,
    };
}
exports.OrderActivityBidToJSON = OrderActivityBidToJSON;
