// tezos-api-client/build/models/OrderActivityList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderActivityListToJSON = exports.OrderActivityListFromJSONTyped = exports.OrderActivityListFromJSON = exports.OrderActivityListTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var OrderActivityListTypeEnum;
(function (OrderActivityListTypeEnum) {
    OrderActivityListTypeEnum["List"] = "list";
})(OrderActivityListTypeEnum = exports.OrderActivityListTypeEnum || (exports.OrderActivityListTypeEnum = {}));
function OrderActivityListFromJSON(json) {
    return OrderActivityListFromJSONTyped(json, false);
}
exports.OrderActivityListFromJSON = OrderActivityListFromJSON;
function OrderActivityListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'hash': json['hash'],
        'maker': json['maker'],
        'make': (0, _1.AssetFromJSON)(json['make']),
        'take': (0, _1.AssetFromJSON)(json['take']),
        'price': json['price'],
    };
}
exports.OrderActivityListFromJSONTyped = OrderActivityListFromJSONTyped;
function OrderActivityListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'hash': value.hash,
        'maker': value.maker,
        'make': (0, _1.AssetToJSON)(value.make),
        'take': (0, _1.AssetToJSON)(value.take),
        'price': value.price,
    };
}
exports.OrderActivityListToJSON = OrderActivityListToJSON;
