// tezos-api-client/build/models/OrderError.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderErrorToJSON = exports.OrderErrorFromJSONTyped = exports.OrderErrorFromJSON = exports.OrderErrorCodeEnum = void 0;
/**
* @export
* @enum {string}
*/
var OrderErrorCodeEnum;
(function (OrderErrorCodeEnum) {
    OrderErrorCodeEnum["OrderInvalidUpdate"] = "ORDER_INVALID_UPDATE";
    OrderErrorCodeEnum["OrderCanceled"] = "ORDER_CANCELED";
    OrderErrorCodeEnum["IncorrectSignature"] = "INCORRECT_SIGNATURE";
})(OrderErrorCodeEnum = exports.OrderErrorCodeEnum || (exports.OrderErrorCodeEnum = {}));
function OrderErrorFromJSON(json) {
    return OrderErrorFromJSONTyped(json, false);
}
exports.OrderErrorFromJSON = OrderErrorFromJSON;
function OrderErrorFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': json['code'],
        'message': json['message'],
    };
}
exports.OrderErrorFromJSONTyped = OrderErrorFromJSONTyped;
function OrderErrorToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
    };
}
exports.OrderErrorToJSON = OrderErrorToJSON;
