// tezos-api-client/build/models/ServerError.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerErrorToJSON = exports.ServerErrorFromJSONTyped = exports.ServerErrorFromJSON = exports.ServerErrorCodeEnum = void 0;
/**
* @export
* @enum {string}
*/
var ServerErrorCodeEnum;
(function (ServerErrorCodeEnum) {
    ServerErrorCodeEnum["UnexpectedApiError"] = "UNEXPECTED_API_ERROR";
})(ServerErrorCodeEnum = exports.ServerErrorCodeEnum || (exports.ServerErrorCodeEnum = {}));
function ServerErrorFromJSON(json) {
    return ServerErrorFromJSONTyped(json, false);
}
exports.ServerErrorFromJSON = ServerErrorFromJSON;
function ServerErrorFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': json['code'],
        'message': json['message'],
    };
}
exports.ServerErrorFromJSONTyped = ServerErrorFromJSONTyped;
function ServerErrorToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
    };
}
exports.ServerErrorToJSON = ServerErrorToJSON;
