// tezos-api-client/build/models/Transfer.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferToJSON = exports.TransferFromJSONTyped = exports.TransferFromJSON = exports.TransferTypeEnum = void 0;
var _1 = require("./");
/**
* @export
* @enum {string}
*/
var TransferTypeEnum;
(function (TransferTypeEnum) {
    TransferTypeEnum["Transfer"] = "transfer";
})(TransferTypeEnum = exports.TransferTypeEnum || (exports.TransferTypeEnum = {}));
function TransferFromJSON(json) {
    return TransferFromJSONTyped(json, false);
}
exports.TransferFromJSON = TransferFromJSON;
function TransferFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['@type'],
        'elt': (0, _1.NftActivityEltFromJSON)(json['elt']),
        'from': json['from'],
    };
}
exports.TransferFromJSONTyped = TransferFromJSONTyped;
function TransferToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        '@type': value.type,
        'elt': (0, _1.NftActivityEltToJSON)(value.elt),
        'from': value.from,
    };
}
exports.TransferToJSON = TransferToJSON;
