// tezos-api-client/build/models/XTZAssetType.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Tezos Rarible API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XTZAssetTypeToJSON = exports.XTZAssetTypeFromJSONTyped = exports.XTZAssetTypeFromJSON = exports.XTZAssetTypeAssetClassEnum = void 0;
/**
* @export
* @enum {string}
*/
var XTZAssetTypeAssetClassEnum;
(function (XTZAssetTypeAssetClassEnum) {
    XTZAssetTypeAssetClassEnum["Xtz"] = "XTZ";
})(XTZAssetTypeAssetClassEnum = exports.XTZAssetTypeAssetClassEnum || (exports.XTZAssetTypeAssetClassEnum = {}));
function XTZAssetTypeFromJSON(json) {
    return XTZAssetTypeFromJSONTyped(json, false);
}
exports.XTZAssetTypeFromJSON = XTZAssetTypeFromJSON;
function XTZAssetTypeFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'assetClass': json['assetClass'],
    };
}
exports.XTZAssetTypeFromJSONTyped = XTZAssetTypeFromJSONTyped;
function XTZAssetTypeToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'assetClass': value.assetClass,
    };
}
exports.XTZAssetTypeToJSON = XTZAssetTypeToJSON;
