// @rarible/sdk/build/sdk-blockchains/tezos/bid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosBid = void 0;
var tslib_1 = require("tslib");
var action_1 = require("@rarible/action");
// eslint-disable-next-line camelcase
var tezos_sdk_1 = require("@rarible/tezos-sdk");
var bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
var types_1 = require("@rarible/types");
// eslint-disable-next-line camelcase
var main_1 = require("@rarible/tezos-sdk/dist/main");
var domain_1 = require("../../types/order/fill/domain");
var retry_1 = require("../../common/retry");
var get_convertable_value_1 = require("../../common/get-convertable-value");
var get_currency_asset_type_1 = require("../../common/get-currency-asset-type");
var common_1 = require("./common");
var TezosBid = /** @class */ (function () {
    function TezosBid(provider, apis, balanceService, network) {
        this.provider = provider;
        this.apis = apis;
        this.balanceService = balanceService;
        this.network = network;
        this.bid = this.bid.bind(this);
        this.update = this.update.bind(this);
        this.getConvertableValue = this.getConvertableValue.bind(this);
        this.convertCurrency = this.convertCurrency.bind(this);
    }
    TezosBid.prototype.getMakeAssetType = function (type) {
        switch (type["@type"]) {
            case "XTZ": {
                return {
                    asset_class: type["@type"],
                };
            }
            case "TEZOS_FT": {
                return {
                    asset_class: "FT",
                    contract: (0, common_1.convertFromContractAddress)(type.contract),
                    token_id: type.tokenId !== undefined ? new bignumber_js_1.default(type.tokenId) : undefined,
                };
            }
            default: {
                throw new Error("Unsupported take asset type");
            }
        }
    };
    TezosBid.prototype.getConvertMap = function () {
        var convertMap = {};
        if (this.provider.config.wrapper) {
            convertMap[(0, common_1.convertTezosToContractAddress)(this.provider.config.wrapper)] = "XTZ";
        }
        return convertMap;
    };
    TezosBid.prototype.getConvertableValue = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var convertMap, assetType, originFeesSum;
            return tslib_1.__generator(this, function (_a) {
                convertMap = this.getConvertMap();
                if (request.assetType) {
                    assetType = request.assetType;
                }
                else if (request.currencyId) {
                    assetType = (0, get_currency_asset_type_1.getCurrencyAssetType)(request.currencyId);
                }
                else {
                    throw new Error("assetType or currencyId should be specified");
                }
                if (assetType["@type"] === "TEZOS_FT" && assetType.contract in convertMap) {
                    originFeesSum = request.originFees.reduce(function (acc, fee) { return fee.value; }, 0);
                    return [2 /*return*/, this.getConvertableValueCommon(assetType, request.price, request.amount, originFeesSum)];
                }
                return [2 /*return*/, undefined];
            });
        });
    };
    TezosBid.prototype.getConvertableValueCommon = function (assetType, price, amount, originFeesSum) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var provider, baseFee, completeFee, value, valueWithFee, _a, _b, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        provider = (0, common_1.getRequiredProvider)(this.provider);
                        baseFee = parseInt(this.provider.config.fees.toString());
                        completeFee = originFeesSum + baseFee;
                        value = new bignumber_js_1.default(price).multipliedBy(amount);
                        return [4 /*yield*/, (0, tezos_sdk_1.add_fee)(provider, { asset_type: (0, common_1.getTezosAssetType)(assetType), value: new bignumber_js_1.default(value) }, new bignumber_js_1.default(completeFee))];
                    case 1:
                        valueWithFee = _d.sent();
                        _a = get_convertable_value_1.getCommonConvertableValue;
                        _b = [this.balanceService.getBalance];
                        _c = common_1.convertTezosToUnionAddress;
                        return [4 /*yield*/, provider.tezos.address()];
                    case 2: return [2 /*return*/, _a.apply(void 0, _b.concat([_c.apply(void 0, [_d.sent()]),
                            valueWithFee.value,
                            { "@type": "XTZ" },
                            assetType]))];
                }
            });
        });
    };
    TezosBid.prototype.convertCurrency = function (convertableValue) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var provider, tx;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = (0, common_1.getRequiredProvider)(this.provider);
                        if (convertableValue === undefined) {
                            return [2 /*return*/];
                        }
                        if (convertableValue.type === "insufficient") {
                            throw new Error("Insufficient XTZ funds");
                        }
                        if (!(convertableValue.type === "convertable")) return [3 /*break*/, 3];
                        return [4 /*yield*/, (0, tezos_sdk_1.wrap)(provider, new bignumber_js_1.default(convertableValue.value))];
                    case 1:
                        tx = _a.sent();
                        return [4 /*yield*/, tx.confirmation()];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    TezosBid.prototype.getWXTZContractAddress = function () {
        var convertMap = this.getConvertMap();
        var wXTZAddressEntry = Object.entries(convertMap)
            .find(function (_a) {
            var _b = tslib_1.__read(_a, 2), contractAddr = _b[0], currency = _b[1];
            return contractAddr && currency === "XTZ";
        });
        if (!wXTZAddressEntry) {
            throw new Error("wXTZ address has not been found");
        }
        var _a = tslib_1.__read(wXTZAddressEntry, 1), wXTZUnionContract = _a[0];
        return (0, types_1.toContractAddress)(wXTZUnionContract);
    };
    TezosBid.prototype.bid = function (prepare) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, itemId, contract, item, itemCollection, submit;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if ("collectionId" in prepare) {
                            throw new Error("Bid collection is not supported");
                        }
                        _a = (0, common_1.getTezosItemData)(prepare.itemId), itemId = _a.itemId, contract = _a.contract;
                        return [4 /*yield*/, (0, retry_1.retry)(90, 1000, function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                return tslib_1.__generator(this, function (_a) {
                                    return [2 /*return*/, this.apis.item.getNftItemById({ itemId: itemId })];
                                });
                            }); })];
                    case 1:
                        item = _b.sent();
                        return [4 /*yield*/, this.apis.collection.getNftCollectionById({
                                collection: contract,
                            })];
                    case 2:
                        itemCollection = _b.sent();
                        submit = action_1.Action.create({
                            id: "convert",
                            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var wXTZUnionAddress, requestCurrency, originFeesSum, value;
                                var _a;
                                return tslib_1.__generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            wXTZUnionAddress = this.getWXTZContractAddress();
                                            requestCurrency = (0, get_currency_asset_type_1.getCurrencyAssetType)(request.currency);
                                            if (!(requestCurrency["@type"] === "TEZOS_FT" && requestCurrency.contract.toLowerCase() === wXTZUnionAddress.toLowerCase())) return [3 /*break*/, 3];
                                            originFeesSum = ((_a = request.originFees) === null || _a === void 0 ? void 0 : _a.reduce(function (acc, fee) { return fee.value; }, 0)) || 0;
                                            return [4 /*yield*/, this.getConvertableValueCommon(requestCurrency, request.price, request.amount, originFeesSum)];
                                        case 1:
                                            value = _b.sent();
                                            return [4 /*yield*/, this.convertCurrency(value)];
                                        case 2:
                                            _b.sent();
                                            _b.label = 3;
                                        case 3: return [2 /*return*/, request];
                                    }
                                });
                            }); },
                        })
                            .thenStep({
                            id: "send-tx",
                            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var provider, makerPublicKey, requestCurrency, order, _a, _b;
                                var _c;
                                return tslib_1.__generator(this, function (_d) {
                                    switch (_d.label) {
                                        case 0:
                                            provider = (0, common_1.getRequiredProvider)(this.provider);
                                            return [4 /*yield*/, (0, common_1.getMakerPublicKey)(provider)];
                                        case 1:
                                            makerPublicKey = _d.sent();
                                            requestCurrency = (0, get_currency_asset_type_1.getCurrencyAssetType)(request.currency);
                                            _a = tezos_sdk_1.bid;
                                            _b = [provider];
                                            _c = {
                                                maker: (0, main_1.pk_to_pkh)(makerPublicKey),
                                                maker_edpk: makerPublicKey,
                                                make_asset_type: this.getMakeAssetType(requestCurrency),
                                                amount: new bignumber_js_1.default(request.amount),
                                                take_asset_type: {
                                                    asset_class: itemCollection.type,
                                                    contract: item.contract,
                                                    token_id: new bignumber_js_1.default(item.tokenId),
                                                },
                                                price: new bignumber_js_1.default(request.price)
                                            };
                                            return [4 /*yield*/, (0, common_1.getPayouts)(provider, request.payouts)];
                                        case 2: return [4 /*yield*/, _a.apply(void 0, _b.concat([(_c.payouts = _d.sent(),
                                                    _c.origin_fees = (0, common_1.convertOrderPayout)(request.originFees),
                                                    _c)]))];
                                        case 3:
                                            order = _d.sent();
                                            return [2 /*return*/, (0, common_1.convertTezosOrderId)(order.hash)];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                multiple: itemCollection.type === "MT",
                                maxAmount: (0, types_1.toBigNumber)(item.supply),
                                originFeeSupport: domain_1.OriginFeeSupport.FULL,
                                payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                                supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                                baseFee: parseInt(this.provider.config.fees.toString()),
                                getConvertableValue: this.getConvertableValue,
                                supportsExpirationDate: false,
                                submit: submit,
                            }];
                }
            });
        });
    };
    TezosBid.prototype.update = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var orderId, order, updateAction;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        orderId = (0, common_1.getTezosOrderId)(request.orderId);
                        return [4 /*yield*/, this.apis.order.getOrderByHash({ hash: orderId })];
                    case 1:
                        order = _a.sent();
                        if (!order) {
                            throw new Error("Order has not been found");
                        }
                        updateAction = action_1.Action.create({
                            id: "convert",
                            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var wXTZUnionAddress, originFeesSum, value;
                                var _a;
                                return tslib_1.__generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            wXTZUnionAddress = (0, common_1.convertFromContractAddress)(this.getWXTZContractAddress());
                                            if (!(order.make.assetType.assetClass === "FT" && order.make.assetType.contract.toLowerCase() === wXTZUnionAddress.toLowerCase())) return [3 /*break*/, 3];
                                            originFeesSum = ((_a = order.data.originFees) === null || _a === void 0 ? void 0 : _a.reduce(function (acc, fee) { return fee.value; }, 0)) || 0;
                                            return [4 /*yield*/, this.getConvertableValueCommon((0, common_1.convertTezosToUnionAsset)(order.make.assetType), request.price, parseInt(order.take.value), originFeesSum)];
                                        case 1:
                                            value = _b.sent();
                                            return [4 /*yield*/, this.convertCurrency(value)];
                                        case 2:
                                            _b.sent();
                                            _b.label = 3;
                                        case 3: return [2 /*return*/, request];
                                    }
                                });
                            }); },
                        })
                            .thenStep({
                            id: "send-tx",
                            run: function (updateRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var provider, orderForm, updatedOrder;
                                var _a, _b;
                                return tslib_1.__generator(this, function (_c) {
                                    switch (_c.label) {
                                        case 0:
                                            provider = (0, common_1.getRequiredProvider)(this.provider);
                                            orderForm = {
                                                type: "RARIBLE_V2",
                                                maker: order.maker,
                                                maker_edpk: order.makerEdpk,
                                                taker_edpk: order.takerEdpk,
                                                make: tslib_1.__assign(tslib_1.__assign({}, (0, common_1.covertToLibAsset)(order.make)), { value: new bignumber_js_1.default(updateRequest.price).multipliedBy(order.take.value) }),
                                                take: (0, common_1.covertToLibAsset)(order.take),
                                                salt: order.salt,
                                                start: order.start,
                                                end: order.end,
                                                signature: order.signature,
                                                data: {
                                                    data_type: "V1",
                                                    payouts: ((_a = order.data.payouts) === null || _a === void 0 ? void 0 : _a.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                    origin_fees: ((_b = order.data.originFees) === null || _b === void 0 ? void 0 : _b.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                },
                                            };
                                            return [4 /*yield*/, (0, tezos_sdk_1.upsert_order)(provider, orderForm, true, true)];
                                        case 1:
                                            updatedOrder = _c.sent();
                                            return [2 /*return*/, (0, common_1.convertTezosOrderId)(updatedOrder.hash)];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                originFeeSupport: domain_1.OriginFeeSupport.FULL,
                                payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                                supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                                baseFee: parseInt(this.provider.config.fees.toString()),
                                getConvertableValue: this.getConvertableValue,
                                submit: updateAction,
                            }];
                }
            });
        });
    };
    return TezosBid;
}());
exports.TezosBid = TezosBid;
