// @rarible/sdk/build/sdk-blockchains/tezos/sell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosSell = void 0;
var tslib_1 = require("tslib");
var sell_1 = require("@rarible/tezos-sdk/dist/order/sell");
// eslint-disable-next-line camelcase
var tezos_sdk_1 = require("@rarible/tezos-sdk");
var action_1 = require("@rarible/action");
var bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
var domain_1 = require("../../types/order/fill/domain");
var retry_1 = require("../../common/retry");
var get_currency_asset_type_1 = require("../../common/get-currency-asset-type");
var common_1 = require("./common");
var TezosSell = /** @class */ (function () {
    function TezosSell(provider, apis) {
        this.provider = provider;
        this.apis = apis;
        this.sell = this.sell.bind(this);
        this.update = this.update.bind(this);
    }
    TezosSell.prototype.parseTakeAssetType = function (type) {
        switch (type["@type"]) {
            case "XTZ":
                return {
                    asset_class: type["@type"],
                };
            case "TEZOS_FT":
                return {
                    asset_class: "FT",
                    contract: (0, common_1.convertFromContractAddress)(type.contract),
                    token_id: type.tokenId ? new bignumber_js_1.default(type.tokenId) : undefined,
                };
            default: {
                throw new Error("Unsupported take asset type");
            }
        }
    };
    TezosSell.prototype.sell = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var submit;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                submit = action_1.Action.create({
                    id: "send-tx",
                    run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                        var provider, makerPublicKey, _a, itemId, contract, item, requestCurrency, itemCollection, tezosRequest, sellOrder;
                        var _b;
                        var _this = this;
                        return tslib_1.__generator(this, function (_c) {
                            switch (_c.label) {
                                case 0:
                                    provider = (0, common_1.getRequiredProvider)(this.provider);
                                    return [4 /*yield*/, (0, common_1.getMakerPublicKey)(provider)];
                                case 1:
                                    makerPublicKey = _c.sent();
                                    _a = (0, common_1.getTezosItemData)(request.itemId), itemId = _a.itemId, contract = _a.contract;
                                    return [4 /*yield*/, (0, retry_1.retry)(90, 1000, function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                            return tslib_1.__generator(this, function (_a) {
                                                return [2 /*return*/, this.apis.item.getNftItemById({ itemId: itemId })];
                                            });
                                        }); })];
                                case 2:
                                    item = _c.sent();
                                    requestCurrency = (0, get_currency_asset_type_1.getCurrencyAssetType)(request.currency);
                                    return [4 /*yield*/, this.apis.collection.getNftCollectionById({
                                            collection: contract,
                                        })];
                                case 3:
                                    itemCollection = _c.sent();
                                    _b = {
                                        maker: (0, tezos_sdk_1.pk_to_pkh)(makerPublicKey),
                                        maker_edpk: makerPublicKey,
                                        make_asset_type: {
                                            asset_class: itemCollection.type,
                                            contract: item.contract,
                                            token_id: new bignumber_js_1.default(item.tokenId),
                                        },
                                        take_asset_type: this.parseTakeAssetType(requestCurrency),
                                        amount: new bignumber_js_1.default(request.amount),
                                        price: new bignumber_js_1.default(request.price)
                                    };
                                    return [4 /*yield*/, (0, common_1.getPayouts)(provider, request.payouts)];
                                case 4:
                                    tezosRequest = (_b.payouts = _c.sent(),
                                        _b.origin_fees = (0, common_1.convertOrderPayout)(request.originFees),
                                        _b);
                                    return [4 /*yield*/, (0, sell_1.sell)(provider, tezosRequest)];
                                case 5:
                                    sellOrder = _c.sent();
                                    return [2 /*return*/, (0, common_1.convertTezosOrderId)(sellOrder.hash)];
                            }
                        });
                    }); },
                });
                return [2 /*return*/, {
                        originFeeSupport: domain_1.OriginFeeSupport.FULL,
                        payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                        supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                        baseFee: parseInt(this.provider.config.fees.toString()),
                        supportsExpirationDate: false,
                        submit: submit,
                    }];
            });
        });
    };
    TezosSell.prototype.update = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var orderId, order, updateAction;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        orderId = (0, common_1.getTezosOrderId)(request.orderId);
                        return [4 /*yield*/, this.apis.order.getOrderByHash({ hash: orderId })];
                    case 1:
                        order = _a.sent();
                        if (!order) {
                            throw new Error("Order has not been found");
                        }
                        updateAction = action_1.Action.create({
                            id: "send-tx",
                            run: function (updateRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var provider, orderForm, updatedOrder;
                                var _a, _b;
                                return tslib_1.__generator(this, function (_c) {
                                    switch (_c.label) {
                                        case 0:
                                            provider = (0, common_1.getRequiredProvider)(this.provider);
                                            orderForm = {
                                                type: "RARIBLE_V2",
                                                maker: order.maker,
                                                maker_edpk: order.makerEdpk,
                                                taker_edpk: order.takerEdpk,
                                                make: (0, common_1.covertToLibAsset)(order.make),
                                                take: tslib_1.__assign(tslib_1.__assign({}, (0, common_1.covertToLibAsset)(order.take)), { value: new bignumber_js_1.default(updateRequest.price).multipliedBy(order.make.value) }),
                                                salt: order.salt,
                                                start: order.start,
                                                end: order.end,
                                                signature: order.signature,
                                                data: {
                                                    data_type: "V1",
                                                    payouts: ((_a = order.data.payouts) === null || _a === void 0 ? void 0 : _a.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                    origin_fees: ((_b = order.data.originFees) === null || _b === void 0 ? void 0 : _b.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                },
                                            };
                                            return [4 /*yield*/, (0, tezos_sdk_1.upsert_order)(provider, orderForm, true, true)];
                                        case 1:
                                            updatedOrder = _c.sent();
                                            return [2 /*return*/, (0, common_1.convertTezosOrderId)(updatedOrder.hash)];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                originFeeSupport: domain_1.OriginFeeSupport.FULL,
                                payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                                supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                                baseFee: parseInt(this.provider.config.fees.toString()),
                                submit: updateAction,
                            }];
                }
            });
        });
    };
    return TezosSell;
}());
exports.TezosSell = TezosSell;
