// @rarible/sdk/build/sdk-blockchains/tezos/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosTransfer = void 0;
var tslib_1 = require("tslib");
var tezos_sdk_1 = require("@rarible/tezos-sdk");
var action_1 = require("@rarible/action");
var types_1 = require("@rarible/types");
var bn_1 = require("@rarible/utils/build/bn");
var sdk_transaction_1 = require("@rarible/sdk-transaction");
var common_1 = require("./common");
var TezosTransfer = /** @class */ (function () {
    function TezosTransfer(provider, apis, network) {
        this.provider = provider;
        this.apis = apis;
        this.network = network;
        this.transfer = this.transfer.bind(this);
    }
    TezosTransfer.prototype.getRequiredProvider = function () {
        if (!(0, common_1.isExistedTezosProvider)(this.provider)) {
            throw new Error("Tezos provider is required");
        }
        return this.provider;
    };
    TezosTransfer.prototype.transfer = function (prepare) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, itemId, contract, item, collection;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (0, common_1.getTezosItemData)(prepare.itemId), itemId = _a.itemId, contract = _a.contract;
                        return [4 /*yield*/, this.apis.item.getNftItemById({ itemId: itemId })];
                    case 1:
                        item = _b.sent();
                        return [4 /*yield*/, this.apis.collection.getNftCollectionById({
                                collection: contract,
                            })];
                    case 2:
                        collection = _b.sent();
                        return [2 /*return*/, {
                                multiple: collection.type === "MT",
                                maxAmount: (0, types_1.toBigNumber)(item.supply),
                                submit: action_1.Action.create({
                                    id: "transfer",
                                    run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                        var amount, result;
                                        return tslib_1.__generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    amount = collection.type === "NFT" ? undefined : (0, bn_1.toBn)((request.amount || 1).toFixed());
                                                    return [4 /*yield*/, (0, tezos_sdk_1.transfer)(this.getRequiredProvider(), {
                                                            contract: item.contract,
                                                            token_id: (0, bn_1.toBn)(item.tokenId),
                                                        }, (0, common_1.getTezosAddress)(request.to), amount)];
                                                case 1:
                                                    result = _a.sent();
                                                    return [2 /*return*/, new sdk_transaction_1.BlockchainTezosTransaction(result, this.network)];
                                            }
                                        });
                                    }); },
                                }),
                            }];
                }
            });
        });
    };
    return TezosTransfer;
}());
exports.TezosTransfer = TezosTransfer;
