// @rarible/sdk/build/sdk-blockchains/ethereum/balance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumBalance = void 0;
var tslib_1 = require("tslib");
var sdk_transaction_1 = require("@rarible/sdk-transaction");
var get_currency_asset_type_1 = require("../../common/get-currency-asset-type");
var common_1 = require("./common");
var EthereumBalance = /** @class */ (function () {
    function EthereumBalance(sdk, network) {
        this.sdk = sdk;
        this.network = network;
        this.getBalance = this.getBalance.bind(this);
        this.convert = this.convert.bind(this);
    }
    EthereumBalance.prototype.getBalance = function (address, currency) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var assetType, convertedAssetType, ethAddress;
            return tslib_1.__generator(this, function (_a) {
                assetType = (0, get_currency_asset_type_1.getCurrencyAssetType)(currency);
                convertedAssetType = (0, common_1.convertToEthereumAssetType)(assetType);
                if (convertedAssetType.assetClass !== "ETH" && convertedAssetType.assetClass !== "ERC20") {
                    throw new Error("Unsupported asset type for getting balance");
                }
                ethAddress = (0, common_1.convertToEthereumAddress)(address);
                return [2 /*return*/, this.sdk.balances.getBalance(ethAddress, convertedAssetType)];
            });
        });
    };
    EthereumBalance.prototype.convert = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var wethContract, from, to, tx;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        wethContract = this.sdk.balances.getWethContractAddress();
                        if (request.isWrap) {
                            from = { assetClass: "ETH" };
                            to = {
                                assetClass: "ERC20",
                                contract: wethContract,
                            };
                        }
                        else {
                            from = {
                                assetClass: "ERC20",
                                contract: wethContract,
                            };
                            to = { assetClass: "ETH" };
                        }
                        return [4 /*yield*/, this.sdk.balances.convert(from, to, request.value)];
                    case 1:
                        tx = _a.sent();
                        return [2 /*return*/, new sdk_transaction_1.BlockchainEthereumTransaction(tx, this.network)];
                }
            });
        });
    };
    return EthereumBalance;
}());
exports.EthereumBalance = EthereumBalance;
