// @rarible/protocol-ethereum-sdk/build/common/apis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEthereumApis = void 0;
var tslib_1 = require("tslib");
var EthereumApiClient = (0, tslib_1.__importStar)(require("@rarible/ethereum-api-client"));
var api_config_1 = require("../config/api-config");
function createEthereumApis(env, params) {
    if (params === void 0) { params = {}; }
    var config = (0, api_config_1.getApiConfig)(env, params);
    var configuration = new EthereumApiClient.Configuration(config);
    return {
        nftItem: new EthereumApiClient.NftItemControllerApi(configuration),
        nftOwnership: new EthereumApiClient.NftOwnershipControllerApi(configuration),
        order: new EthereumApiClient.OrderControllerApi(configuration),
        orderActivity: new EthereumApiClient.OrderActivityControllerApi(configuration),
        nftCollection: new EthereumApiClient.NftCollectionControllerApi(configuration),
        balances: new EthereumApiClient.BalanceControllerApi(configuration),
        gateway: new EthereumApiClient.GatewayControllerApi(configuration),
        nftLazyMint: new EthereumApiClient.NftLazyMintControllerApi(configuration),
        auction: new EthereumApiClient.AuctionControllerApi(configuration),
    };
}
exports.createEthereumApis = createEthereumApis;
