// @rarible/protocol-ethereum-sdk/build/common/balances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Balances = void 0;
var tslib_1 = require("tslib");
var utils_1 = require("@rarible/utils");
var Balances = /** @class */ (function () {
    function Balances(apis) {
        this.apis = apis;
        this.getBalance = this.getBalance.bind(this);
    }
    Balances.prototype.getBalance = function (address, assetType) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, ethBalance, balance;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = assetType.assetClass;
                        switch (_a) {
                            case "ETH": return [3 /*break*/, 1];
                            case "ERC20": return [3 /*break*/, 3];
                        }
                        return [3 /*break*/, 5];
                    case 1: return [4 /*yield*/, this.apis.balances.getEthBalance({ owner: address })];
                    case 2:
                        ethBalance = _b.sent();
                        return [2 /*return*/, (0, utils_1.toBn)(ethBalance.decimalBalance)];
                    case 3: return [4 /*yield*/, this.apis.balances.getErc20Balance({
                            contract: assetType.contract,
                            owner: address,
                        })];
                    case 4:
                        balance = _b.sent();
                        return [2 /*return*/, (0, utils_1.toBn)(balance.decimalBalance)];
                    case 5: throw new Error("Asset class is not supported");
                }
            });
        });
    };
    return Balances;
}());
exports.Balances = Balances;
