// @rarible/protocol-ethereum-sdk/build/common/mint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createErc1155V2Collection = exports.createErc1155V1Collection = exports.createErc721V1Collection = exports.createErc721V3Collection = exports.createErc721V2Collection = void 0;
var ethereum_api_client_1 = require("@rarible/ethereum-api-client");
var domain_1 = require("../nft/contracts/domain");
function createErc721V2Collection(address) {
    return {
        features: [ethereum_api_client_1.NftCollectionFeatures.SECONDARY_SALE_FEES],
        id: address,
        name: "Test-collection",
        type: ethereum_api_client_1.NftCollectionType.ERC721,
        version: domain_1.ERC721VersionEnum.ERC721V2,
    };
}
exports.createErc721V2Collection = createErc721V2Collection;
function createErc721V3Collection(address) {
    return {
        features: [ethereum_api_client_1.NftCollectionFeatures.SECONDARY_SALE_FEES, ethereum_api_client_1.NftCollectionFeatures.MINT_AND_TRANSFER],
        id: address,
        name: "Test-collection",
        type: ethereum_api_client_1.NftCollectionType.ERC721,
        version: domain_1.ERC721VersionEnum.ERC721V3,
    };
}
exports.createErc721V3Collection = createErc721V3Collection;
function createErc721V1Collection(address) {
    return {
        features: [],
        id: address,
        name: "Test-collection",
        type: ethereum_api_client_1.NftCollectionType.ERC721,
        version: domain_1.ERC721VersionEnum.ERC721V1,
    };
}
exports.createErc721V1Collection = createErc721V1Collection;
function createErc1155V1Collection(address) {
    return {
        features: [ethereum_api_client_1.NftCollectionFeatures.SECONDARY_SALE_FEES],
        id: address,
        name: "Test-collection",
        type: ethereum_api_client_1.NftCollectionType.ERC1155,
        version: domain_1.ERC1155VersionEnum.ERC1155V1,
    };
}
exports.createErc1155V1Collection = createErc1155V1Collection;
function createErc1155V2Collection(address) {
    return {
        features: [ethereum_api_client_1.NftCollectionFeatures.MINT_AND_TRANSFER],
        id: address,
        name: "Test-collection",
        type: ethereum_api_client_1.NftCollectionType.ERC1155,
        version: domain_1.ERC1155VersionEnum.ERC1155V2,
    };
}
exports.createErc1155V2Collection = createErc1155V2Collection;
