// @rarible/sdk/build/sdk-blockchains/solana/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaCollection = void 0;
var tslib_1 = require("tslib");
var action_1 = require("@rarible/action");
var types_1 = require("@rarible/types");
var api_client_1 = require("@rarible/api-client");
var sdk_transaction_1 = require("@rarible/sdk-transaction");
var SolanaCollection = /** @class */ (function () {
    function SolanaCollection(sdk, wallet, apis, config) {
        var _this = this;
        this.sdk = sdk;
        this.wallet = wallet;
        this.apis = apis;
        this.config = config;
        this.createCollection = action_1.Action.create({
            id: "send-tx",
            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var mintPrepare, res;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (request.blockchain !== api_client_1.Blockchain.SOLANA) {
                                throw new Error("Wrong blockchain");
                            }
                            return [4 /*yield*/, this.sdk.nft.mint({
                                    metadataUrl: request.asset.arguments.metadataURI,
                                    signer: this.wallet.provider,
                                    maxSupply: 0,
                                    collection: null,
                                })];
                        case 1:
                            mintPrepare = _a.sent();
                            return [4 /*yield*/, mintPrepare.tx.submit("single")];
                        case 2:
                            res = _a.sent();
                            return [2 /*return*/, {
                                    tx: new sdk_transaction_1.BlockchainSolanaTransaction(res, this.sdk),
                                    address: (0, types_1.toContractAddress)("SOLANA:".concat(mintPrepare.mint.toString())),
                                }];
                    }
                });
            }); },
        });
    }
    return SolanaCollection;
}());
exports.SolanaCollection = SolanaCollection;
