// @rarible/sdk/build/sdk-blockchains/solana/fill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaFill = void 0;
var tslib_1 = require("tslib");
var action_1 = require("@rarible/action");
var api_client_1 = require("@rarible/api-client");
var sdk_transaction_1 = require("@rarible/sdk-transaction");
var domain_1 = require("../../types/order/fill/domain");
var address_converters_1 = require("./common/address-converters");
var order_1 = require("./common/order");
var auction_house_1 = require("./common/auction-house");
var SolanaFill = /** @class */ (function () {
    function SolanaFill(sdk, wallet, apis, config) {
        this.sdk = sdk;
        this.wallet = wallet;
        this.apis = apis;
        this.config = config;
        this.fill = this.fill.bind(this);
    }
    SolanaFill.prototype.isBuyOrder = function (order) {
        return order.make.type["@type"] === "SOLANA_NFT";
    };
    SolanaFill.prototype.fill = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var order;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet) {
                            throw new Error("Solana wallet not provided");
                        }
                        return [4 /*yield*/, (0, order_1.getPreparedOrder)(request, this.apis)];
                    case 1:
                        order = _a.sent();
                        if (order.status !== api_client_1.OrderStatus.ACTIVE) {
                            throw new Error("Order is not active");
                        }
                        return [2 /*return*/, this.isBuyOrder(order) ? this.buy(order) : this.acceptBid(order)];
                }
            });
        });
    };
    SolanaFill.prototype.buy = function (order) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var auctionHouse, mint, price, item, submit;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        auctionHouse = (0, address_converters_1.extractPublicKey)((0, order_1.getOrderData)(order).auctionHouse);
                        mint = (0, order_1.getMintId)(order);
                        price = (0, order_1.getPrice)(order);
                        return [4 /*yield*/, this.apis.item.getItemById({ itemId: (0, order_1.getItemId)(mint) })];
                    case 1:
                        item = _a.sent();
                        submit = action_1.Action
                            .create({
                            id: "send-tx",
                            run: function (buyRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var buyPrepare, executePrepare;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.sdk.order.buy({
                                                auctionHouse: auctionHouse,
                                                signer: this.wallet.provider,
                                                mint: mint,
                                                price: price,
                                                tokensAmount: buyRequest.amount,
                                            })];
                                        case 1:
                                            buyPrepare = _a.sent();
                                            return [4 /*yield*/, this.sdk.order.executeSell({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    buyerWallet: this.wallet.provider.publicKey,
                                                    sellerWallet: (0, address_converters_1.extractPublicKey)(order.maker),
                                                    mint: mint,
                                                    price: price,
                                                    tokensAmount: buyRequest.amount,
                                                })];
                                        case 2:
                                            executePrepare = _a.sent();
                                            return [2 /*return*/, this.sdk.unionInstructionsAndSend(this.wallet.provider, [buyPrepare, executePrepare], "processed")];
                                    }
                                });
                            }); },
                        })
                            .after(function (tx) { return new sdk_transaction_1.BlockchainSolanaTransaction(tx, _this.sdk); });
                        return [2 /*return*/, {
                                multiple: parseFloat(item.supply.toString()) > 1,
                                maxAmount: order.makeStock,
                                baseFee: 0,
                                supportsPartialFill: true,
                                originFeeSupport: domain_1.OriginFeeSupport.NONE,
                                payoutsSupport: domain_1.PayoutsSupport.NONE,
                                submit: submit,
                            }];
                }
            });
        });
    };
    SolanaFill.prototype.acceptBid = function (order) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var auctionHouse, mint, price, item, submit;
            var _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        auctionHouse = (0, address_converters_1.extractPublicKey)((0, order_1.getOrderData)(order).auctionHouse);
                        mint = (0, order_1.getMintId)(order);
                        price = (0, order_1.getPrice)(order);
                        return [4 /*yield*/, this.apis.item.getItemById({ itemId: (0, order_1.getItemId)(mint) })];
                    case 1:
                        item = _b.sent();
                        submit = action_1.Action
                            .create({
                            id: "send-tx",
                            run: function (buyRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var sellPrepare, executePrepare;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.sdk.order.sell({
                                                auctionHouse: auctionHouse,
                                                signer: this.wallet.provider,
                                                mint: mint,
                                                price: price,
                                                tokensAmount: buyRequest.amount,
                                            })];
                                        case 1:
                                            sellPrepare = _a.sent();
                                            return [4 /*yield*/, this.sdk.order.executeSell({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    buyerWallet: (0, address_converters_1.extractPublicKey)(order.maker),
                                                    sellerWallet: this.wallet.provider.publicKey,
                                                    mint: mint,
                                                    price: price,
                                                    tokensAmount: buyRequest.amount,
                                                })];
                                        case 2:
                                            executePrepare = _a.sent();
                                            return [2 /*return*/, this.sdk.unionInstructionsAndSend(this.wallet.provider, [sellPrepare, executePrepare], "processed")];
                                    }
                                });
                            }); },
                        })
                            .after(function (tx) { return new sdk_transaction_1.BlockchainSolanaTransaction(tx, _this.sdk); });
                        _a = {
                            multiple: parseFloat(item.supply.toString()) > 1,
                            maxAmount: order.makeStock
                        };
                        return [4 /*yield*/, (0, auction_house_1.getAuctionHouseFee)((0, order_1.getOrderData)(order).auctionHouse)];
                    case 2: return [2 /*return*/, (_a.baseFee = _b.sent(),
                            _a.supportsPartialFill = true,
                            _a.originFeeSupport = domain_1.OriginFeeSupport.NONE,
                            _a.payoutsSupport = domain_1.PayoutsSupport.NONE,
                            _a.submit = submit,
                            _a)];
                }
            });
        });
    };
    return SolanaFill;
}());
exports.SolanaFill = SolanaFill;
