// @rarible/sdk/build/sdk-blockchains/solana/order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaOrder = void 0;
var tslib_1 = require("tslib");
var action_1 = require("@rarible/action");
var types_1 = require("@rarible/types");
var sdk_transaction_1 = require("@rarible/sdk-transaction");
var domain_1 = require("../../types/order/fill/domain");
var auction_house_1 = require("./common/auction-house");
var address_converters_1 = require("./common/address-converters");
var order_1 = require("./common/order");
var currencies_1 = require("./common/currencies");
var WRAPPED_SOL = "So11111111111111111111111111111111111111112";
var SolanaOrder = /** @class */ (function () {
    function SolanaOrder(sdk, wallet, apis, config) {
        var _this = this;
        this.sdk = sdk;
        this.wallet = wallet;
        this.apis = apis;
        this.config = config;
        this.cancel = action_1.Action.create({
            id: "send-tx",
            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var order, orderData, res;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, (0, order_1.getPreparedOrder)(request, this.apis)];
                        case 1:
                            order = _a.sent();
                            orderData = (0, order_1.getOrderData)(order);
                            return [4 /*yield*/, this.sdk.order.cancel({
                                    auctionHouse: (0, address_converters_1.extractPublicKey)(orderData.auctionHouse),
                                    signer: this.wallet.provider,
                                    mint: (0, order_1.getMintId)(order),
                                    price: (0, order_1.getPrice)(order),
                                    tokensAmount: (0, order_1.getTokensAmount)(order),
                                })];
                        case 2: return [4 /*yield*/, (_a.sent()).submit("processed")];
                        case 3:
                            res = _a.sent();
                            return [2 /*return*/, new sdk_transaction_1.BlockchainSolanaTransaction(res, this.sdk)];
                    }
                });
            }); },
        });
        this.sell = this.sell.bind(this);
        this.bid = this.bid.bind(this);
        this.sellUpdate = this.sellUpdate.bind(this);
        this.bidUpdate = this.bidUpdate.bind(this);
    }
    SolanaOrder.prototype.sell = function () {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var auctionHouse, submit;
            var _b;
            var _this = this;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!this.wallet) {
                            throw new Error("Solana wallet not provided");
                        }
                        auctionHouse = (0, auction_house_1.getAuctionHouse)({ "@type": "SOLANA_SOL" }, (_a = this.config) === null || _a === void 0 ? void 0 : _a.auctionHouseMapping);
                        submit = action_1.Action.create({
                            id: "send-tx",
                            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var mint;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            mint = (0, address_converters_1.extractPublicKey)(request.itemId);
                                            return [4 /*yield*/, this.sdk.order.sell({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    mint: mint,
                                                    price: parseFloat(request.price.toString()),
                                                    tokensAmount: request.amount,
                                                })];
                                        case 1: return [4 /*yield*/, (_a.sent()).submit("processed")];
                                        case 2:
                                            _a.sent();
                                            return [2 /*return*/, (0, order_1.getOrderId)("SELL", this.wallet.provider.publicKey.toString(), mint.toString(), auctionHouse.toString())];
                                    }
                                });
                            }); },
                        });
                        _b = {
                            originFeeSupport: domain_1.OriginFeeSupport.NONE,
                            payoutsSupport: domain_1.PayoutsSupport.NONE,
                            supportedCurrencies: (0, currencies_1.getCurrencies)()
                        };
                        return [4 /*yield*/, (0, auction_house_1.getAuctionHouseFee)(auctionHouse)];
                    case 1: return [2 /*return*/, (_b.baseFee = _c.sent(),
                            _b.supportsExpirationDate = false,
                            _b.submit = submit,
                            _b)];
                }
            });
        });
    };
    SolanaOrder.prototype.sellUpdate = function (prepareRequest) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var order, updateAction;
            var _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet) {
                            throw new Error("Solana wallet not provided");
                        }
                        return [4 /*yield*/, (0, order_1.getPreparedOrder)(prepareRequest, this.apis)];
                    case 1:
                        order = _b.sent();
                        updateAction = action_1.Action.create({
                            id: "send-tx",
                            run: function (updateRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var mint, auctionHouse;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            mint = (0, order_1.getMintId)(order);
                                            auctionHouse = (0, address_converters_1.extractPublicKey)((0, order_1.getOrderData)(order).auctionHouse);
                                            return [4 /*yield*/, this.sdk.order.sell({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    mint: mint,
                                                    price: parseFloat(updateRequest.price.toString()),
                                                    tokensAmount: (0, order_1.getTokensAmount)(order),
                                                })];
                                        case 1: return [4 /*yield*/, (_a.sent()).submit("processed")];
                                        case 2:
                                            _a.sent();
                                            return [2 /*return*/, (0, order_1.getOrderId)("SELL", this.wallet.provider.publicKey.toString(), mint.toString(), auctionHouse.toString())];
                                    }
                                });
                            }); },
                        });
                        _a = {
                            originFeeSupport: domain_1.OriginFeeSupport.NONE,
                            payoutsSupport: domain_1.PayoutsSupport.NONE,
                            supportedCurrencies: (0, currencies_1.getCurrencies)()
                        };
                        return [4 /*yield*/, (0, auction_house_1.getAuctionHouseFee)((0, order_1.getOrderData)(order).auctionHouse)];
                    case 2: return [2 /*return*/, (_a.baseFee = _b.sent(),
                            _a.submit = updateAction,
                            _a)];
                }
            });
        });
    };
    SolanaOrder.prototype.getConvertableValue = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, undefined];
            });
        });
    };
    SolanaOrder.prototype.bid = function (prepare) {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var auctionHouse, item, submit;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet) {
                            throw new Error("Solana wallet not provided");
                        }
                        if (!("itemId" in prepare)) {
                            throw new Error("No ItemId provided");
                        }
                        auctionHouse = (0, auction_house_1.getAuctionHouse)({ "@type": "SOLANA_SOL" }, (_a = this.config) === null || _a === void 0 ? void 0 : _a.auctionHouseMapping);
                        return [4 /*yield*/, this.apis.item.getItemById({ itemId: prepare.itemId })];
                    case 1:
                        item = _b.sent();
                        submit = action_1.Action.create({
                            id: "send-tx",
                            run: function (request) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var mint;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            mint = (0, address_converters_1.extractPublicKey)(prepare.itemId);
                                            return [4 /*yield*/, this.sdk.order.buy({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    mint: mint,
                                                    price: parseFloat(request.price.toString()),
                                                    tokensAmount: request.amount,
                                                })];
                                        case 1: return [4 /*yield*/, (_a.sent()).submit("processed")];
                                        case 2:
                                            _a.sent();
                                            return [2 /*return*/, (0, order_1.getOrderId)("BUY", this.wallet.provider.publicKey.toString(), WRAPPED_SOL, //todo should be ah currency program address
                                                auctionHouse.toString())];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                multiple: parseFloat(item.supply) > 1,
                                maxAmount: (0, types_1.toBigNumber)(item.supply),
                                originFeeSupport: domain_1.OriginFeeSupport.NONE,
                                payoutsSupport: domain_1.PayoutsSupport.NONE,
                                supportedCurrencies: (0, currencies_1.getCurrencies)(),
                                baseFee: 0,
                                getConvertableValue: this.getConvertableValue,
                                supportsExpirationDate: false,
                                submit: submit,
                            }];
                }
            });
        });
    };
    SolanaOrder.prototype.bidUpdate = function (prepareRequest) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var order, updateAction;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet) {
                            throw new Error("Solana wallet not provided");
                        }
                        return [4 /*yield*/, (0, order_1.getPreparedOrder)(prepareRequest, this.apis)];
                    case 1:
                        order = _a.sent();
                        updateAction = action_1.Action.create({
                            id: "send-tx",
                            run: function (updateRequest) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var mint, auctionHouse;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            mint = (0, order_1.getMintId)(order);
                                            auctionHouse = (0, address_converters_1.extractPublicKey)((0, order_1.getOrderData)(order).auctionHouse);
                                            return [4 /*yield*/, this.sdk.order.buy({
                                                    auctionHouse: auctionHouse,
                                                    signer: this.wallet.provider,
                                                    mint: mint,
                                                    price: parseFloat(updateRequest.price.toString()),
                                                    tokensAmount: (0, order_1.getTokensAmount)(order),
                                                })];
                                        case 1: return [4 /*yield*/, (_a.sent()).submit("processed")];
                                        case 2:
                                            _a.sent();
                                            return [2 /*return*/, (0, order_1.getOrderId)("BUY", this.wallet.provider.publicKey.toString(), WRAPPED_SOL, //todo should be ah currency program address
                                                auctionHouse.toString())];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                originFeeSupport: domain_1.OriginFeeSupport.NONE,
                                payoutsSupport: domain_1.PayoutsSupport.NONE,
                                supportedCurrencies: (0, currencies_1.getCurrencies)(),
                                baseFee: 0,
                                getConvertableValue: this.getConvertableValue,
                                submit: updateAction,
                            }];
                }
            });
        });
    };
    return SolanaOrder;
}());
exports.SolanaOrder = SolanaOrder;
