// @rarible/protocol-ethereum-sdk/build/order/fill-order/open-sea-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderOpenSeaV1DataV1HowToCall = exports.OrderOpenSeaV1DataV1SaleKind = exports.OrderOpenSeaV1DataV1Side = exports.OrderOpenSeaV1DataV1FeeMethod = void 0;
var OrderOpenSeaV1DataV1FeeMethod;
(function (OrderOpenSeaV1DataV1FeeMethod) {
    OrderOpenSeaV1DataV1FeeMethod[OrderOpenSeaV1DataV1FeeMethod["PROTOCOL_FEE"] = 0] = "PROTOCOL_FEE";
    OrderOpenSeaV1DataV1FeeMethod[OrderOpenSeaV1DataV1FeeMethod["SPLIT_FEE"] = 1] = "SPLIT_FEE";
})(OrderOpenSeaV1DataV1FeeMethod = exports.OrderOpenSeaV1DataV1FeeMethod || (exports.OrderOpenSeaV1DataV1FeeMethod = {}));
var OrderOpenSeaV1DataV1Side;
(function (OrderOpenSeaV1DataV1Side) {
    OrderOpenSeaV1DataV1Side[OrderOpenSeaV1DataV1Side["BUY"] = 0] = "BUY";
    OrderOpenSeaV1DataV1Side[OrderOpenSeaV1DataV1Side["SELL"] = 1] = "SELL";
})(OrderOpenSeaV1DataV1Side = exports.OrderOpenSeaV1DataV1Side || (exports.OrderOpenSeaV1DataV1Side = {}));
var OrderOpenSeaV1DataV1SaleKind;
(function (OrderOpenSeaV1DataV1SaleKind) {
    OrderOpenSeaV1DataV1SaleKind[OrderOpenSeaV1DataV1SaleKind["FIXED_PRICE"] = 0] = "FIXED_PRICE";
    OrderOpenSeaV1DataV1SaleKind[OrderOpenSeaV1DataV1SaleKind["DUTCH_AUCTION"] = 1] = "DUTCH_AUCTION";
})(OrderOpenSeaV1DataV1SaleKind = exports.OrderOpenSeaV1DataV1SaleKind || (exports.OrderOpenSeaV1DataV1SaleKind = {}));
var OrderOpenSeaV1DataV1HowToCall;
(function (OrderOpenSeaV1DataV1HowToCall) {
    OrderOpenSeaV1DataV1HowToCall[OrderOpenSeaV1DataV1HowToCall["CALL"] = 0] = "CALL";
    OrderOpenSeaV1DataV1HowToCall[OrderOpenSeaV1DataV1HowToCall["DELEGATE_CALL"] = 1] = "DELEGATE_CALL";
})(OrderOpenSeaV1DataV1HowToCall = exports.OrderOpenSeaV1DataV1HowToCall || (exports.OrderOpenSeaV1DataV1HowToCall = {}));
