// @rarible/protocol-ethereum-sdk/build/order/fill-order/rarible-v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RaribleV2OrderHandler = void 0;
var tslib_1 = require("tslib");
var types_1 = require("@rarible/types");
var sign_order_1 = require("../sign-order");
var get_asset_with_fee_1 = require("../get-asset-with-fee");
var approve_1 = require("../approve");
var exchange_v2_1 = require("../contracts/exchange-v2");
var wait_tx_1 = require("../../common/wait-tx");
var is_signer_1 = require("../../common/is-signer");
var fix_signature_1 = require("../../common/fix-signature");
var invert_order_1 = require("./invert-order");
var RaribleV2OrderHandler = /** @class */ (function () {
    function RaribleV2OrderHandler(ethereum, send, config, getBaseOrderFeeConfig) {
        this.ethereum = ethereum;
        this.send = send;
        this.config = config;
        this.getBaseOrderFeeConfig = getBaseOrderFeeConfig;
    }
    RaribleV2OrderHandler.prototype.invert = function (request, maker) {
        var inverted = (0, invert_order_1.invertOrder)(request.order, request.amount, maker);
        switch (request.order.data.dataType) {
            case "RARIBLE_V2_DATA_V1": {
                inverted.data = {
                    dataType: "RARIBLE_V2_DATA_V1",
                    originFees: request.originFees || [],
                    payouts: request.payouts || [],
                };
                break;
            }
            case "RARIBLE_V2_DATA_V2": {
                inverted.data = {
                    dataType: "RARIBLE_V2_DATA_V2",
                    originFees: request.originFees || [],
                    payouts: request.payouts || [],
                    isMakeFill: !request.order.data.isMakeFill,
                };
                break;
            }
            default: throw new Error("Unsupported order dataType");
        }
        return inverted;
    };
    RaribleV2OrderHandler.prototype.approve = function (order, infinite) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var withFee;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.ethereum) {
                            throw new Error("Wallet undefined");
                        }
                        return [4 /*yield*/, this.getMakeAssetWithFee(order)];
                    case 1:
                        withFee = _a.sent();
                        return [4 /*yield*/, (0, wait_tx_1.waitTx)((0, approve_1.approve)(this.ethereum, this.send, this.config.transferProxies, order.maker, withFee, infinite))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.getTransactionData = function (initial, inverted) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var exchangeContract, functionCall, _a, _b, _c;
            var _d;
            return (0, tslib_1.__generator)(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (!this.ethereum) {
                            throw new Error("Wallet undefined");
                        }
                        exchangeContract = (0, exchange_v2_1.createExchangeV2Contract)(this.ethereum, this.config.exchange.v2);
                        _b = (_a = exchangeContract).functionCall;
                        _c = ["matchOrders"];
                        return [4 /*yield*/, this.fixForTx(initial)];
                    case 1:
                        functionCall = _b.apply(_a, _c.concat([_e.sent(), (0, fix_signature_1.fixSignature)(initial.signature) || "0x",
                            (0, sign_order_1.orderToStruct)(this.ethereum, inverted),
                            (0, fix_signature_1.fixSignature)(inverted.signature) || "0x"]));
                        _d = {
                            functionCall: functionCall
                        };
                        return [4 /*yield*/, this.getMatchV2Options(initial, inverted)];
                    case 2: return [2 /*return*/, (_d.options = _e.sent(),
                            _d)];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.sendTransaction = function (initial, inverted) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, functionCall, options;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getTransactionData(initial, inverted)];
                    case 1:
                        _a = _b.sent(), functionCall = _a.functionCall, options = _a.options;
                        return [2 /*return*/, this.send(functionCall, options)];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.fixForTx = function (order) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var hash, isMakerSigner;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.ethereum) {
                            throw new Error("Wallet undefined");
                        }
                        hash = (0, sign_order_1.hashToSign)(this.config, this.ethereum, order);
                        return [4 /*yield*/, (0, is_signer_1.isSigner)(this.ethereum, order.maker, hash, order.signature)];
                    case 1:
                        isMakerSigner = _a.sent();
                        return [2 /*return*/, (0, sign_order_1.orderToStruct)(this.ethereum, order, !isMakerSigner)];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.getMatchV2Options = function (left, right) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var asset, asset;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(left.make.assetType.assetClass === "ETH" && left.salt === types_1.ZERO_WORD)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.getMakeAssetWithFee(left)];
                    case 1:
                        asset = _a.sent();
                        return [2 /*return*/, { value: asset.value }];
                    case 2:
                        if (!(right.make.assetType.assetClass === "ETH" && right.salt === types_1.ZERO_WORD)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.getMakeAssetWithFee(right)];
                    case 3:
                        asset = _a.sent();
                        return [2 /*return*/, { value: asset.value }];
                    case 4: return [2 /*return*/, {}];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.getMakeAssetWithFee = function (order) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, _b;
            return (0, tslib_1.__generator)(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = get_asset_with_fee_1.getAssetWithFee;
                        _b = [order.make];
                        return [4 /*yield*/, this.getOrderFee(order)];
                    case 1: return [2 /*return*/, _a.apply(void 0, _b.concat([_c.sent()]))];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.getOrderFee = function (order) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = order.data.originFees.map(function (f) { return f.value; }).reduce(function (v, acc) { return v + acc; }, 0);
                        return [4 /*yield*/, this.getBaseOrderFee()];
                    case 1: return [2 /*return*/, _a + (_b.sent())];
                }
            });
        });
    };
    RaribleV2OrderHandler.prototype.getBaseOrderFee = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, this.getBaseOrderFeeConfig("RARIBLE_V2")];
            });
        });
    };
    return RaribleV2OrderHandler;
}());
exports.RaribleV2OrderHandler = RaribleV2OrderHandler;
