// @rarible/solana-sdk/build/sdk/collection/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaCollectionSdk = void 0;
var tslib_1 = require("tslib");
var prepared_transaction_1 = require("../prepared-transaction");
var verify_collection_1 = require("./verifyCollection/verify-collection");
var SolanaCollectionSdk = /** @class */ (function () {
    function SolanaCollectionSdk(connection, logger) {
        this.connection = connection;
        this.logger = logger;
    }
    SolanaCollectionSdk.prototype.verifyCollection = function (request) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var instructions;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, verify_collection_1.getVerifyCollectionInstructions)({
                            connection: this.connection,
                            signer: request.signer,
                            mint: request.mint,
                            collection: request.collection,
                        })];
                    case 1:
                        instructions = _a.sent();
                        return [2 /*return*/, new prepared_transaction_1.PreparedTransaction(this.connection, instructions, request.signer, this.logger, function () {
                                _this.logger.log("Mint", request.mint.toString(), "approved to collection", request.collection.toString());
                            })];
                }
            });
        });
    };
    return SolanaCollectionSdk;
}());
exports.SolanaCollectionSdk = SolanaCollectionSdk;
