// @rarible/protocol-ethereum-sdk/build/config/e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.e2eConfig = void 0;
var types_1 = require("@rarible/types");
var id_1 = require("../common/id");
var common_1 = require("./common");
exports.e2eConfig = {
    basePath: "https://ethereum-api-e2e.rarible.org",
    chainId: 17,
    exchange: {
        v1: (0, types_1.toAddress)("0x80f32a12cc4c095e2a409b70e5c96e8515e87dea"),
        v2: (0, types_1.toAddress)("0x551E4009116d489e3C5a98405A9c4B601D250B58"),
        openseaV1: types_1.ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toAddress)("0x66611f8d97688a0af08d4337d7846efec6995d58"),
        erc20: (0, types_1.toAddress)("0xbf558e78cfde95afbf17a4abe394cb2cc42e6270"),
        erc721Lazy: (0, types_1.toAddress)("0xe853B9994304264ff418b818A8D23FD39e8DABe6"),
        erc1155Lazy: (0, types_1.toAddress)("0x6E605A7d1FD15e9087f0756ab57E0ED99735a7a7"),
        openseaV1: types_1.ZERO_ADDRESS,
        cryptoPunks: (0, types_1.toAddress)("0x3268d39B77A7F189209E87E13eedA613098b32c7"),
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toAddress)("0x74B4721599213fdC7eD0bb67D3B25a6fb3c74B61"),
        erc1155: (0, types_1.toAddress)("0xF35BC450544C7d8c2559B4DAeb51E9617F39e7C6"),
    },
    weth: (0, types_1.toAddress)("0xc6f33b62a94939e52e1b074c4ac1a801b869fdb2"),
    auction: (0, types_1.toAddress)("0xE091e5511815107072ed1C42d0d00f151D6197ed"),
};
