// @rarible/protocol-ethereum-sdk/build/config/mumbai-dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mumbaiDevConfig = void 0;
var types_1 = require("@rarible/types");
var id_1 = require("../common/id");
var common_1 = require("./common");
exports.mumbaiDevConfig = {
    basePath: "https://polygon-api-dev.rarible.org",
    chainId: 80001,
    exchange: {
        v1: types_1.ZERO_ADDRESS,
        v2: (0, types_1.toAddress)("0x2Fc743F5419637B93dDAC159715B902186300041"),
        openseaV1: types_1.ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toAddress)("0x02e21199D043dab90248f79d6A8d0c36832734B0"),
        erc20: (0, types_1.toAddress)("0xCA90c2E7d3c41EF220888B0038849BA1e67688bC"),
        erc721Lazy: (0, types_1.toAddress)("0xA8628124C255171cF356F0E0204E2D19CA89F636"),
        erc1155Lazy: (0, types_1.toAddress)("0x8fd27f771bf72C1F8c8b78FCdcf323C0C34f030b"),
        openseaV1: types_1.ZERO_ADDRESS,
        cryptoPunks: types_1.ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toAddress)("0xa85180a21786bA65b0778bE1cb5CBA5E5c6cD21d"),
        erc1155: (0, types_1.toAddress)("0xAa9CD5834E0009902EeAA3FEfAc6A160e9A096b4"),
    },
    weth: (0, types_1.toAddress)("0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa"),
    auction: types_1.ZERO_ADDRESS,
};
