// @rarible/protocol-ethereum-sdk/build/config/polygon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.polygonConfig = void 0;
var types_1 = require("@rarible/types");
var id_1 = require("../common/id");
var common_1 = require("./common");
exports.polygonConfig = {
    basePath: "https://polygon-api.rarible.org",
    chainId: 137,
    exchange: {
        v1: types_1.ZERO_ADDRESS,
        v2: (0, types_1.toAddress)("0x12b3897a36fDB436ddE2788C06Eff0ffD997066e"),
        openseaV1: types_1.ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toAddress)("0xd47e14DD9b98411754f722B4c4074e14752Ada7C"),
        erc20: (0, types_1.toAddress)("0x49b4e47079d9b733B2227fa15f0762dBF707B263"),
        erc721Lazy: (0, types_1.toAddress)("0xDD28328257a2Cce3204332C747Cc350153937A1D"),
        erc1155Lazy: (0, types_1.toAddress)("0x0E63021A7597B254484b7F99dDD9b319591350B6"),
        openseaV1: types_1.ZERO_ADDRESS,
        cryptoPunks: types_1.ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toAddress)("0x16911a36a56f828f17632cD4915614Dd5c7a45e0"),
        erc1155: (0, types_1.toAddress)("0xF46e8e6fA0F048DdD76F8c6982eBD059796298B8"),
    },
    weth: (0, types_1.toAddress)("0x7ceB23fD6bC0adD59E62ac25578270cFf1b9f619"),
    auction: types_1.ZERO_ADDRESS,
};
