// @rarible/protocol-ethereum-sdk/build/config/polygon-dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.devPolygonConfig = void 0;
var types_1 = require("@rarible/types");
var id_1 = require("../common/id");
var common_1 = require("./common");
exports.devPolygonConfig = {
    basePath: "https://dev-polygon-api.rarible.org",
    chainId: 300501,
    exchange: {
        v1: types_1.ZERO_ADDRESS,
        v2: (0, types_1.toAddress)("0x8283Ffd0F535E1103C3599D2d00b85815774A896"),
        openseaV1: types_1.ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toAddress)("0x3586d3E6CDaE98d5F0eEaB737977Bc78406Da2BD"),
        erc20: (0, types_1.toAddress)("0xeC47DA9591FC24F5a5F401e8D275526Cc5eE5d37"),
        erc721Lazy: (0, types_1.toAddress)("0x8812cFb55853da0968a02AaaEA84CD93EC4b42A1"),
        erc1155Lazy: (0, types_1.toAddress)("0x87ECcc03BaBC550c919Ad61187Ab597E9E7f7C21"),
        openseaV1: types_1.ZERO_ADDRESS,
        cryptoPunks: (0, types_1.toAddress)("0x44a72AEb7dAc73c4b72f89d6855dE063949627F3"),
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toAddress)("0x66611f8D97688A0aF08D4337D7846eFEc6995d58"),
        erc1155: (0, types_1.toAddress)("0x31C827f06E10e4999eb88c193669d408eF597B3D"),
    },
    weth: (0, types_1.toAddress)("0xb24740adECB71fEb7d66AA4EBD5F5c334E5bE922"),
    auction: types_1.ZERO_ADDRESS,
};
