// @rarible/protocol-ethereum-sdk/build/config/ropsten.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ropstenConfig = void 0;
var types_1 = require("@rarible/types");
var id_1 = require("../common/id");
var common_1 = require("./common");
exports.ropstenConfig = {
    basePath: "https://ethereum-api-dev.rarible.org",
    chainId: 3,
    exchange: {
        v1: (0, types_1.toAddress)("0xd782A10D023828d283f7b943Ae0fc3F07B2952d9"),
        v2: (0, types_1.toAddress)("0x33Aef288C093Bf7b36fBe15c3190e616a993b0AD"),
        openseaV1: (0, types_1.toAddress)("0x5206e78b21ce315ce284fb24cf05e0585a93b1d9"),
    },
    transferProxies: {
        nft: (0, types_1.toAddress)("0xf8e4ecac18b65fd04569ff1f0d561f74effaa206"),
        erc20: (0, types_1.toAddress)("0xa5a51d7b4933185da9c932e5375187f661cb0c69"),
        erc721Lazy: (0, types_1.toAddress)("0x6c49c170c82C40709a32Fb4E827ad3011CD86227"),
        erc1155Lazy: (0, types_1.toAddress)("0x9F7fBc52A53f85e57a5DAde35dFa14797A4dA412"),
        openseaV1: types_1.ZERO_ADDRESS,
        cryptoPunks: (0, types_1.toAddress)("0x6B8ABca06F4D08310dc09540B8E27E5A0D9572E9"),
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toAddress)("0x939d0308CE4274C287E7305D381B336B77dBfcd3"),
        erc1155: (0, types_1.toAddress)("0xccf0cB91Fe5cCb697781427C141ed0662aE4FE2e"),
    },
    weth: (0, types_1.toAddress)("0xc778417e063141139fce010982780140aa0cd5ab"),
    auction: types_1.ZERO_ADDRESS,
};
