// @rarible/protocol-ethereum-sdk/build/nft/prepare-mint-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareMintRequest = exports.isMultiple = exports.isSupportsLazyMint = exports.isSupportsRoyalties = void 0;
var mint_1 = require("./mint");
function isSupportsRoyalties(collection) {
    if (collection.type === "ERC721") {
        return (0, mint_1.isErc721v3Collection)(collection) || (0, mint_1.isErc721v2Collection)(collection);
    }
    else if (collection.type === "ERC1155") {
        return true;
    }
    else {
        throw new Error("Unrecognized collection type");
    }
}
exports.isSupportsRoyalties = isSupportsRoyalties;
function isSupportsLazyMint(collection) {
    return (0, mint_1.isErc721v3Collection)(collection) || (0, mint_1.isErc1155v2Collection)(collection);
}
exports.isSupportsLazyMint = isSupportsLazyMint;
function isMultiple(collection) {
    return collection.type === "ERC1155";
}
exports.isMultiple = isMultiple;
function prepareMintRequest(collection) {
    return {
        multiple: isMultiple(collection),
        supportsRoyalties: isSupportsRoyalties(collection),
        supportsLazyMint: isSupportsLazyMint(collection),
    };
}
exports.prepareMintRequest = prepareMintRequest;
