// @rarible/logger/build/union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnionLogger = void 0;
var UnionLogger = /** @class */ (function () {
    function UnionLogger(loggers) {
        this.loggers = loggers;
    }
    UnionLogger.prototype.debug = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.loggers.forEach(function (l) { return l.debug.apply(l, args); });
    };
    UnionLogger.prototype.error = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.loggers.forEach(function (l) { return l.error.apply(l, args); });
    };
    UnionLogger.prototype.info = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.loggers.forEach(function (l) { return l.info.apply(l, args); });
    };
    UnionLogger.prototype.trace = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.loggers.forEach(function (l) { return l.trace.apply(l, args); });
    };
    UnionLogger.prototype.warn = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.loggers.forEach(function (l) { return l.warn.apply(l, args); });
    };
    return UnionLogger;
}());
exports.UnionLogger = UnionLogger;
